<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Exploratory Data Analysis of PhIP-Seq data from Shrock 2020</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="2024-06-19_Shrock2020_eda_files/libs/clipboard/clipboard.min.js"></script>
<script src="2024-06-19_Shrock2020_eda_files/libs/quarto-html/quarto.js"></script>
<script src="2024-06-19_Shrock2020_eda_files/libs/quarto-html/popper.min.js"></script>
<script src="2024-06-19_Shrock2020_eda_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="2024-06-19_Shrock2020_eda_files/libs/quarto-html/anchor.min.js"></script>
<link href="2024-06-19_Shrock2020_eda_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="2024-06-19_Shrock2020_eda_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="2024-06-19_Shrock2020_eda_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="2024-06-19_Shrock2020_eda_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="2024-06-19_Shrock2020_eda_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exploratory Data Analysis of PhIP-Seq data from Shrock 2020</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<ul>
<li>this is a good representation of publicly available PhIP-seq data</li>
<li>link to paper: https://www.science.org/doi/full/10.1126/science.abd4250</li>
<li>for detailed annotations of datasets, see HDSI Grant – 2024-2027/Published PhIP-seq data/VirScan - Shrock, Science, 2020/Shrock_2020_annotations.md on SharePoint</li>
</ul>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.stats <span class="im">import</span> ttest_ind</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="co"># five distinct dataframes here</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">'../data/Shrock_2020/'</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>pat_meta <span class="op">=</span> pd.read_csv(data_dir <span class="op">+</span> <span class="st">'Shrock_2020_patient-metadata.csv'</span>, encoding<span class="op">=</span><span class="st">'utf-8'</span>, engine<span class="op">=</span><span class="st">'python'</span>, on_bad_lines<span class="op">=</span><span class="st">'warn'</span>) </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>cor_lib <span class="op">=</span> pd.read_csv(data_dir <span class="op">+</span> <span class="st">'Shrock_2020_Coronavirus-library.csv'</span>, encoding<span class="op">=</span><span class="st">'utf-8'</span>, engine<span class="op">=</span><span class="st">'python'</span>, on_bad_lines<span class="op">=</span><span class="st">'warn'</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>cor_scr_z <span class="op">=</span> pd.read_csv(data_dir <span class="op">+</span> <span class="st">'Shrock_2020_Coronavirus-screen-IgG-Zscores.csv'</span>, encoding<span class="op">=</span><span class="st">'utf-8'</span>, engine<span class="op">=</span><span class="st">'python'</span>, on_bad_lines<span class="op">=</span><span class="st">'warn'</span>) </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>virs_z <span class="op">=</span> pd.read_csv(data_dir <span class="op">+</span> <span class="st">'Shrock_2020_VirScan-IgG-Zscores.csv'</span>, encoding<span class="op">=</span><span class="st">'utf-8'</span>, engine<span class="op">=</span><span class="st">'python'</span>, on_bad_lines<span class="op">=</span><span class="st">'warn'</span>) </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>virs_lib <span class="op">=</span> pd.read_csv(data_dir <span class="op">+</span> <span class="st">'Shrock_2020_VirScan-library.csv'</span>, encoding<span class="op">=</span><span class="st">'utf-8'</span>, engine<span class="op">=</span><span class="st">'python'</span>, on_bad_lines<span class="op">=</span><span class="st">'warn'</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"1. patient-metadata"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>display(pat_meta)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"patient-metadata columns: </span><span class="sc">{</span>pat_meta<span class="sc">.</span>columns<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>1. patient-metadata
patient-metadata columns: Index(['rep_1', 'rep_2', 'Sample', 'Library', 'IP reagent', 'COVID19_status',
       'Hospitalized', 'Pull down antibody'],
      dtype='object')
============================</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">rep_1</th>
<th data-quarto-table-cell-role="th">rep_2</th>
<th data-quarto-table-cell-role="th">Sample</th>
<th data-quarto-table-cell-role="th">Library</th>
<th data-quarto-table-cell-role="th">IP reagent</th>
<th data-quarto-table-cell-role="th">COVID19_status</th>
<th data-quarto-table-cell-role="th">Hospitalized</th>
<th data-quarto-table-cell-role="th">Pull down antibody</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>77_79_S23</td>
<td>78_80_S23</td>
<td>BWH_1</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>positive</td>
<td>non-hospitalized</td>
<td>IgG</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>73_75_S1</td>
<td>74_76_S1</td>
<td>820-0447-1</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>positive</td>
<td>non-hospitalized</td>
<td>IgG</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>73_75_S2</td>
<td>74_76_S2</td>
<td>820-0144-1</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>positive</td>
<td>non-hospitalized</td>
<td>IgG</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>73_75_S3</td>
<td>74_76_S3</td>
<td>820-0084-1</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>positive</td>
<td>non-hospitalized</td>
<td>IgG</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>73_75_S4</td>
<td>74_76_S4</td>
<td>820-0117-1</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>positive</td>
<td>non-hospitalized</td>
<td>IgG</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">417</td>
<td>81_83_S186</td>
<td>82_84_S186</td>
<td>7/25/19_serum_F</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>negative</td>
<td>NaN</td>
<td>IgG</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">418</td>
<td>81_83_S187</td>
<td>82_84_S187</td>
<td>7/25/19_serum_G</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>negative</td>
<td>NaN</td>
<td>IgG</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">419</td>
<td>81_83_S188</td>
<td>82_84_S188</td>
<td>7/25/19_serum_H</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>negative</td>
<td>NaN</td>
<td>IgG</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">420</td>
<td>81_83_S189</td>
<td>82_84_S189</td>
<td>7/25/19_serum_I</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>negative</td>
<td>NaN</td>
<td>IgG</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">421</td>
<td>81_83_S190</td>
<td>82_84_S190</td>
<td>7/25/19_serum_J</td>
<td>vir3 + CoV 56mer + CoV 20mer</td>
<td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>
<td>negative</td>
<td>NaN</td>
<td>IgG</td>
</tr>
</tbody>
</table>

<p>422 rows × 8 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'patient-metadata sanity checks'</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># missing values in the dataset</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==missing data row counts=="</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pat_meta.isnull().<span class="bu">sum</span>())</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># drop rows with missing values</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># data = data.dropna()</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># check data types</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==data types=="</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pat_meta.dtypes)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary statistics for numerical data</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==summary stats=="</span>)</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pat_meta.describe())</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Frequency of categories for COVID-19 status</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==covid19 status=="</span>)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(pat_meta[<span class="st">'COVID19_status'</span>].value_counts())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>patient-metadata sanity checks
==missing data row counts==
rep_1                   0
rep_2                   0
Sample                  0
Library                 0
IP reagent              0
COVID19_status          0
Hospitalized          190
Pull down antibody      0
dtype: int64
============================
==data types==
rep_1                 object
rep_2                 object
Sample                object
Library               object
IP reagent            object
COVID19_status        object
Hospitalized          object
Pull down antibody    object
dtype: object
============================
==summary stats==
             rep_1       rep_2           Sample                       Library  \
count          422         422              422                           422   
unique         422         422              422                             1   
top     81_83_S190  82_84_S190  7/25/19_serum_J  vir3 + CoV 56mer + CoV 20mer   
freq             1           1                1                           422   

                                               IP reagent COVID19_status  \
count                                                 422            422   
unique                                                  1              2   
top     6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   
freq                                                  422            232   

            Hospitalized Pull down antibody  
count                232                422  
unique                 2                  1  
top     non-hospitalized                IgG  
freq                 131                422  
============================
==covid19 status==
COVID19_status
positive    232
negative    190
Name: count, dtype: int64</code></pre>
</div>
</div>
<section id="questions" class="level2">
<h2 class="anchored" data-anchor-id="questions">Questions</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"2. VirScan-library"</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>display(virs_lib)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"VirScan-library columns: </span><span class="sc">{</span>virs_lib<span class="sc">.</span>columns<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>2. VirScan-library
VirScan-library columns: Index(['Unnamed: 0', 'Aclstr50', 'Bclstr50', 'Entry', 'Gene names',
       'Gene ontology (GO)', 'Gene ontology IDs', 'Genus', 'Organism',
       'Protein names', 'Sequence', 'Species', 'Subcellular location',
       'Version (entry)', 'Version (sequence)', 'end', 'id', 'oligo', 'source',
       'start', 'peptide'],
      dtype='object')
============================</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Unnamed: 0</th>
<th data-quarto-table-cell-role="th">Aclstr50</th>
<th data-quarto-table-cell-role="th">Bclstr50</th>
<th data-quarto-table-cell-role="th">Entry</th>
<th data-quarto-table-cell-role="th">Gene names</th>
<th data-quarto-table-cell-role="th">Gene ontology (GO)</th>
<th data-quarto-table-cell-role="th">Gene ontology IDs</th>
<th data-quarto-table-cell-role="th">Genus</th>
<th data-quarto-table-cell-role="th">Organism</th>
<th data-quarto-table-cell-role="th">Protein names</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Species</th>
<th data-quarto-table-cell-role="th">Subcellular location</th>
<th data-quarto-table-cell-role="th">Version (entry)</th>
<th data-quarto-table-cell-role="th">Version (sequence)</th>
<th data-quarto-table-cell-role="th">end</th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">oligo</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">start</th>
<th data-quarto-table-cell-role="th">peptide</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0</td>
<td>8652.0</td>
<td>15383.0</td>
<td>A0A126</td>
<td>US5</td>
<td>suppression by virus of host apoptotic process</td>
<td>GO:0019050</td>
<td>NaN</td>
<td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>
<td>Glycoprotein J</td>
<td>...</td>
<td>Papiine herpesvirus 2</td>
<td>NaN</td>
<td>13.0</td>
<td>1.0</td>
<td>56</td>
<td>1</td>
<td>aGGAATTCCGCTGCGTatgcgcagcttgctgtttgtggtcggtgct...</td>
<td>Vir2</td>
<td>1</td>
<td>MRSLLFVVGAWVAALVTNLTPDAALASGTTTTAAAGNTSATASPGD...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1</td>
<td>15383.0</td>
<td>15384.0</td>
<td>A0A126</td>
<td>US5</td>
<td>suppression by virus of host apoptotic process</td>
<td>GO:0019050</td>
<td>NaN</td>
<td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>
<td>Glycoprotein J</td>
<td>...</td>
<td>Papiine herpesvirus 2</td>
<td>NaN</td>
<td>13.0</td>
<td>1.0</td>
<td>84</td>
<td>2</td>
<td>aTGAATTCGGAGCGGTactacaaccaccgctgccgcagggaacaca...</td>
<td>Vir2</td>
<td>29</td>
<td>TTTTAAAGNTSATASPGDNATSIDAGSTITAAAPPGHSTPWPALPT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>2</td>
<td>15384.0</td>
<td>3254.0</td>
<td>A0A126</td>
<td>US5</td>
<td>suppression by virus of host apoptotic process</td>
<td>GO:0019050</td>
<td>NaN</td>
<td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>
<td>Glycoprotein J</td>
<td>...</td>
<td>Papiine herpesvirus 2</td>
<td>NaN</td>
<td>13.0</td>
<td>1.0</td>
<td>112</td>
<td>3</td>
<td>aGGAATTCCGCTGCGTattaccgctgccgctcctccaggtcattca...</td>
<td>Vir2</td>
<td>57</td>
<td>ITAAAPPGHSTPWPALPTDLALPLVIGGLCALTLAAMGAGALLHRC...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>3</td>
<td>3254.0</td>
<td>NaN</td>
<td>A0A126</td>
<td>US5</td>
<td>suppression by virus of host apoptotic process</td>
<td>GO:0019050</td>
<td>NaN</td>
<td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>
<td>Glycoprotein J</td>
<td>...</td>
<td>Papiine herpesvirus 2</td>
<td>NaN</td>
<td>13.0</td>
<td>1.0</td>
<td>118</td>
<td>4</td>
<td>aTGAATTCGGAGCGGTttgtgcgccctcacactcgcagcaatgggc...</td>
<td>Vir2</td>
<td>85</td>
<td>LCALTLAAMGAGALLHRCCRRCARRRQNVSSVSA</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>4</td>
<td>32531.0</td>
<td>NaN</td>
<td>A0A130</td>
<td>US4</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>
<td>Glycoprotein G (Fragment)</td>
<td>...</td>
<td>Papiine herpesvirus 2</td>
<td>NaN</td>
<td>8.0</td>
<td>1.0</td>
<td>20</td>
<td>5</td>
<td>aGGAATTCCGCTGCGTcgcgatcgcggcccttctcgctctcgcgtg...</td>
<td>Vir2</td>
<td>1</td>
<td>RDRGPSRSRVRYTRLAASEA</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">128252</td>
<td>128283</td>
<td>NaN</td>
<td>NaN</td>
<td>Q2VST0</td>
<td>NaN</td>
<td>hydrolase activity, hydrolyzing O-glycosyl com...</td>
<td>GO:0004553; GO:0005975</td>
<td>Ziziphus</td>
<td>Ziziphus mauritiana</td>
<td>Allergen Ziz m 1</td>
<td>...</td>
<td>Ziziphus mauritiana</td>
<td>NaN</td>
<td>42.0</td>
<td>1.0</td>
<td>224</td>
<td>128283</td>
<td>GGAATTCCGCTGCGTAAAGACTTGACAAAGGATCGCACGCGCCCGT...</td>
<td>IEDB</td>
<td>169</td>
<td>KDLTKDRTRPFYLSAAPKCSAYNDSDAYLWTAVETGLFDFVWVKFY...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">128253</td>
<td>128284</td>
<td>NaN</td>
<td>NaN</td>
<td>Q2VST0</td>
<td>NaN</td>
<td>hydrolase activity, hydrolyzing O-glycosyl com...</td>
<td>GO:0004553; GO:0005975</td>
<td>Ziziphus</td>
<td>Ziziphus mauritiana</td>
<td>Allergen Ziz m 1</td>
<td>...</td>
<td>Ziziphus mauritiana</td>
<td>NaN</td>
<td>42.0</td>
<td>1.0</td>
<td>252</td>
<td>128284</td>
<td>GGAATTCCGCTGCGTTTATGGACTGCGGTCGAGACTGGCCTTTTTG...</td>
<td>IEDB</td>
<td>197</td>
<td>LWTAVETGLFDFVWVKFYNDTSCQYNNDTAAGLDAFYRSWYDWTVS...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">128254</td>
<td>128285</td>
<td>NaN</td>
<td>NaN</td>
<td>Q2VST0</td>
<td>NaN</td>
<td>hydrolase activity, hydrolyzing O-glycosyl com...</td>
<td>GO:0004553; GO:0005975</td>
<td>Ziziphus</td>
<td>Ziziphus mauritiana</td>
<td>Allergen Ziz m 1</td>
<td>...</td>
<td>Ziziphus mauritiana</td>
<td>NaN</td>
<td>42.0</td>
<td>1.0</td>
<td>280</td>
<td>128285</td>
<td>GGAATTCCGCTGCGTACAGCGGCTGGGCTGGATGCCTTCTACCGCA...</td>
<td>IEDB</td>
<td>225</td>
<td>TAAGLDAFYRSWYDWTVSLAEGNKLLIGIPASNETDNSPLGGYIPS...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">128255</td>
<td>128286</td>
<td>NaN</td>
<td>NaN</td>
<td>Q2VST0</td>
<td>NaN</td>
<td>hydrolase activity, hydrolyzing O-glycosyl com...</td>
<td>GO:0004553; GO:0005975</td>
<td>Ziziphus</td>
<td>Ziziphus mauritiana</td>
<td>Allergen Ziz m 1</td>
<td>...</td>
<td>Ziziphus mauritiana</td>
<td>NaN</td>
<td>42.0</td>
<td>1.0</td>
<td>308</td>
<td>128286</td>
<td>GGAATTCCGCTGCGTATCCCTGCCAGCAACGAAACGGATAATAGCC...</td>
<td>IEDB</td>
<td>253</td>
<td>IPASNETDNSPLGGYIPSDVLNDQIVSVIMTSSKFGGVNVWNRYYD...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">128256</td>
<td>128287</td>
<td>NaN</td>
<td>NaN</td>
<td>Q2VST0</td>
<td>NaN</td>
<td>hydrolase activity, hydrolyzing O-glycosyl com...</td>
<td>GO:0004553; GO:0005975</td>
<td>Ziziphus</td>
<td>Ziziphus mauritiana</td>
<td>Allergen Ziz m 1</td>
<td>...</td>
<td>Ziziphus mauritiana</td>
<td>NaN</td>
<td>42.0</td>
<td>1.0</td>
<td>330</td>
<td>128287</td>
<td>GGAATTCCGCTGCGTATTATGACAAGTTCAAAATTTGGCGGGGTCA...</td>
<td>IEDB</td>
<td>281</td>
<td>IMTSSKFGGVNVWNRYYDLKTNYSSSIILEYVNSGTKYLPLRTKFM...</td>
</tr>
</tbody>
</table>

<p>128257 rows × 21 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'VirScan-library sanity checks'</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># missing values in the dataset</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==missing data row counts=="</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(virs_lib.isnull().<span class="bu">sum</span>())</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># check data types</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==data types=="</span>)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(virs_lib.dtypes)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary statistics for numerical data</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==summary stats=="</span>)</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(virs_lib[[<span class="st">'start'</span>, <span class="st">'end'</span>]].describe())</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>VirScan-library sanity checks
==missing data row counts==
Unnamed: 0                   0
Aclstr50                 19655
Bclstr50                 23874
Entry                        0
Gene names               43678
Gene ontology (GO)       19432
Gene ontology IDs        48087
Genus                   121852
Organism                  2187
Protein names                0
Sequence                     0
Species                      0
Subcellular location     75754
Version (entry)           2187
Version (sequence)        2187
end                          0
id                           0
oligo                        0
source                       0
start                        0
peptide                      0
dtype: int64
============================
==data types==
Unnamed: 0                int64
Aclstr50                float64
Bclstr50                float64
Entry                    object
Gene names               object
Gene ontology (GO)       object
Gene ontology IDs        object
Genus                    object
Organism                 object
Protein names            object
Sequence                 object
Species                  object
Subcellular location     object
Version (entry)         float64
Version (sequence)      float64
end                       int64
id                        int64
oligo                    object
source                   object
start                     int64
peptide                  object
dtype: object
============================
==summary stats==
               start            end
count  128257.000000  128257.000000
mean      542.589387     596.482102
std       840.458759     840.761116
min         0.000000       6.000000
25%        85.000000     140.000000
50%       253.000000     308.000000
75%       589.000000     644.000000
max      7197.000000    7241.000000
============================</code></pre>
</div>
</div>
</section>
<section id="questions-1" class="level2">
<h2 class="anchored" data-anchor-id="questions-1">Questions</h2>
<div class="cell" data-tags="[]">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'3. Coronavirus-library'</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>display(cor_lib)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Coronavirus-library columns: </span><span class="sc">{</span>cor_lib<span class="sc">.</span>columns<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3. Coronavirus-library
Coronavirus-library columns: Index(['Unnamed: 0', 'Library type', 'Organsim', 'Barcode ID', 'Peptide ID',
       'Protein ID', 'Full protein sequence', 'Protein length', 'Protein name',
       'Peptide sequence', 'Start position', 'End position', 'Peptide length',
       'Nucleotide sequence'],
      dtype='object')
============================</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Unnamed: 0</th>
<th data-quarto-table-cell-role="th">Library type</th>
<th data-quarto-table-cell-role="th">Organsim</th>
<th data-quarto-table-cell-role="th">Barcode ID</th>
<th data-quarto-table-cell-role="th">Peptide ID</th>
<th data-quarto-table-cell-role="th">Protein ID</th>
<th data-quarto-table-cell-role="th">Full protein sequence</th>
<th data-quarto-table-cell-role="th">Protein length</th>
<th data-quarto-table-cell-role="th">Protein name</th>
<th data-quarto-table-cell-role="th">Peptide sequence</th>
<th data-quarto-table-cell-role="th">Start position</th>
<th data-quarto-table-cell-role="th">End position</th>
<th data-quarto-table-cell-role="th">Peptide length</th>
<th data-quarto-table-cell-role="th">Nucleotide sequence</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0</td>
<td>Coronavirus 56mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>YP_009725301.9.BC1</td>
<td>YP_009725301.9</td>
<td>YP_009725301</td>
<td>SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...</td>
<td>307</td>
<td>3C-like proteinase</td>
<td>PLSAQTGIAVLDMCASLKELLQNGMNGRTILGSALLEDEFTPFDVV...</td>
<td>251</td>
<td>307</td>
<td>55</td>
<td>GAATTCGGAGCGGTCCTTTGTCAGCACAAACAGGCATTGCCGTGCT...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1</td>
<td>Coronavirus 56mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>YP_009725299.68.BC1</td>
<td>YP_009725299.68</td>
<td>YP_009725299</td>
<td>APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...</td>
<td>1946</td>
<td>nsp3</td>
<td>IALIWNVKDFMSLSEQLRKQIRSAAKKNNLPFKLTCATTRQVVNVV...</td>
<td>1890</td>
<td>1946</td>
<td>55</td>
<td>GAATTCGGAGCGGTATTGCTCTTATTTGGAACGTGAAAGACTTTAT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>2</td>
<td>Coronavirus 56mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>YP_009725306.3.BC1</td>
<td>YP_009725306.3</td>
<td>YP_009725306</td>
<td>AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...</td>
<td>140</td>
<td>nsp10</td>
<td>NPKGFCDLKGKYVQIPTTCANDPVGFTLKNTVCTVCGMWKGYGCSC...</td>
<td>84</td>
<td>140</td>
<td>55</td>
<td>GAATTCGGAGCGGTAATCCTAAGGGCTTTTGCGACCTCAAGGGGAA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>3</td>
<td>Coronavirus 56mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>YP_009725312.0.BC1</td>
<td>YP_009725312.0</td>
<td>YP_009725312</td>
<td>SADAQSFLNGFAV*</td>
<td>14</td>
<td>nsp11</td>
<td>SADAQSFLNGFAV*</td>
<td>0</td>
<td>14</td>
<td>13</td>
<td>GAATTCGGAGCGGTTCCGCTGATGCCCAGTCATTTCTGAACGGTTT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>4</td>
<td>Coronavirus 56mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>YP_009725308.20.BC1</td>
<td>YP_009725308.20</td>
<td>YP_009725308</td>
<td>AVGACVLCNSQTSLRCGACIRRPFLCCKCCYDHVISTSHKLVLSVN...</td>
<td>602</td>
<td>helicase</td>
<td>TQTTETAHSCNVNRFNVAITRAKVGILCIMSDRDLYDKLQFTSLEI...</td>
<td>546</td>
<td>602</td>
<td>55</td>
<td>GAATTCGGAGCGGTACTCAGACAACAGAGACAGCACACTCATGCAA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">58773</td>
<td>58773</td>
<td>SARS-CoV-2 20mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>QID21074.20aa.2.BC2</td>
<td>QID21074.20aa.2</td>
<td>QID21074</td>
<td>MKFLVFLGIIITVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>
<td>122</td>
<td>orf8 protein</td>
<td>ITVAAFHQECSLQSCTQHQP*</td>
<td>10</td>
<td>30</td>
<td>20</td>
<td>GAATTCCGCTGCGTATTACAGTGGCCGCCTTTCACCAGGAGTGCTC...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">58774</td>
<td>58774</td>
<td>SARS-CoV-2 20mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>QID21067.20aa.1257.BC2</td>
<td>QID21067.20aa.1257</td>
<td>QID21067</td>
<td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>
<td>7097</td>
<td>orf1ab polyprotein</td>
<td>YKIEELFYSYATHSDKFTAG*</td>
<td>6285</td>
<td>6305</td>
<td>20</td>
<td>GAATTCCGCTGCGTTATAAAATTGAAGAATTGTTTTATTCATACGC...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">58775</td>
<td>58775</td>
<td>SARS-CoV-2 20mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>QID21067.20aa.1258.BC2</td>
<td>QID21067.20aa.1258</td>
<td>QID21067</td>
<td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>
<td>7097</td>
<td>orf1ab polyprotein</td>
<td>LFYSYATHSDKFTAGVCLFW*</td>
<td>6290</td>
<td>6310</td>
<td>20</td>
<td>GAATTCCGCTGCGTTTGTTTTACTCATACGCTACTCATAGTGACAA...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">58776</td>
<td>58776</td>
<td>SARS-CoV-2 20mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>QID21067.20aa.1259.BC2</td>
<td>QID21067.20aa.1259</td>
<td>QID21067</td>
<td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>
<td>7097</td>
<td>orf1ab polyprotein</td>
<td>ATHSDKFTAGVCLFWNCNVD*</td>
<td>6295</td>
<td>6315</td>
<td>20</td>
<td>GAATTCCGCTGCGTGCTACTCACAGCGACAAATTTACAGCCGGTGT...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">58777</td>
<td>58777</td>
<td>SARS-CoV-2 20mer</td>
<td>Severe acute respiratory syndrome coronavirus 2</td>
<td>QID21067.20aa.1260.BC2</td>
<td>QID21067.20aa.1260</td>
<td>QID21067</td>
<td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>
<td>7097</td>
<td>orf1ab polyprotein</td>
<td>KFTAGVCLFWNCNVDRYPAN*</td>
<td>6300</td>
<td>6320</td>
<td>20</td>
<td>GAATTCCGCTGCGTAAGTTCACAGCTGGTGTGTGTCTTTTCTGGAA...</td>
</tr>
</tbody>
</table>

<p>58778 rows × 14 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Coronavirus-library sanity checks'</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># missing values in the dataset</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==missing data row counts=="</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cor_lib.isnull().<span class="bu">sum</span>())</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># check data types</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==data types=="</span>)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cor_lib.dtypes)</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary statistics for numerical data</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==summary stats=="</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cor_lib[[<span class="st">'Protein length'</span>, <span class="st">'Start position'</span>, <span class="st">'End position'</span>, <span class="st">'Peptide length'</span>]].describe())</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Coronavirus-library sanity checks
==missing data row counts==
Unnamed: 0               0
Library type             0
Organsim                 0
Barcode ID               0
Peptide ID               0
Protein ID               0
Full protein sequence    0
Protein length           0
Protein name             0
Peptide sequence         0
Start position           0
End position             0
Peptide length           0
Nucleotide sequence      0
dtype: int64
============================
==data types==
Unnamed: 0                int64
Library type             object
Organsim                 object
Barcode ID               object
Peptide ID               object
Protein ID               object
Full protein sequence    object
Protein length            int64
Protein name             object
Peptide sequence         object
Start position            int64
End position              int64
Peptide length            int64
Nucleotide sequence      object
dtype: object
============================
==summary stats==
       Protein length  Start position  End position  Peptide length
count    58778.000000    58778.000000  58778.000000    58778.000000
mean      5078.358161     2492.860288   2546.176427       53.255980
std       2930.797729     2238.995872   2238.958572        9.412306
min         14.000000        0.000000     14.000000       13.000000
25%       1274.000000      364.000000    420.000000       56.000000
50%       7097.000000     1932.000000   1988.000000       56.000000
75%       7097.000000     4340.000000   4396.000000       56.000000
max       7153.000000     7097.000000   7153.000000       56.000000
============================</code></pre>
</div>
</div>
<div class="cell" data-tags="[]">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'4. VirScan-IgG-Zscores'</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>display(virs_z)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"VirScan-IgG-Zscores columns: </span><span class="sc">{</span>virs_z<span class="sc">.</span>columns<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>4. VirScan-IgG-Zscores
VirScan-IgG-Zscores columns: Index(['id', 'group', '73_75_S1', '73_75_S2', '73_75_S3', '73_75_S4',
       '73_75_S5', '73_75_S6', '73_75_S7', '73_75_S8',
       ...
       '82_84_S182', '82_84_S183', '82_84_S184', '82_84_S185', '82_84_S186',
       '82_84_S187', '82_84_S188', '82_84_S189', '82_84_S190', 'input'],
      dtype='object', length=1101)
============================</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">group</th>
<th data-quarto-table-cell-role="th">73_75_S1</th>
<th data-quarto-table-cell-role="th">73_75_S2</th>
<th data-quarto-table-cell-role="th">73_75_S3</th>
<th data-quarto-table-cell-role="th">73_75_S4</th>
<th data-quarto-table-cell-role="th">73_75_S5</th>
<th data-quarto-table-cell-role="th">73_75_S6</th>
<th data-quarto-table-cell-role="th">73_75_S7</th>
<th data-quarto-table-cell-role="th">73_75_S8</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">82_84_S182</th>
<th data-quarto-table-cell-role="th">82_84_S183</th>
<th data-quarto-table-cell-role="th">82_84_S184</th>
<th data-quarto-table-cell-role="th">82_84_S185</th>
<th data-quarto-table-cell-role="th">82_84_S186</th>
<th data-quarto-table-cell-role="th">82_84_S187</th>
<th data-quarto-table-cell-role="th">82_84_S188</th>
<th data-quarto-table-cell-role="th">82_84_S189</th>
<th data-quarto-table-cell-role="th">82_84_S190</th>
<th data-quarto-table-cell-role="th">input</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>1</td>
<td>10</td>
<td>0.57</td>
<td>-0.95</td>
<td>-0.84</td>
<td>2.74</td>
<td>-0.78</td>
<td>0.70</td>
<td>-0.73</td>
<td>-0.83</td>
<td>...</td>
<td>-0.81</td>
<td>0.61</td>
<td>-0.78</td>
<td>-0.65</td>
<td>1.02</td>
<td>-0.65</td>
<td>-0.57</td>
<td>1.34</td>
<td>-0.78</td>
<td>9</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>2</td>
<td>269</td>
<td>-1.43</td>
<td>0.79</td>
<td>1.35</td>
<td>0.75</td>
<td>2.16</td>
<td>2.66</td>
<td>2.72</td>
<td>-0.48</td>
<td>...</td>
<td>-0.24</td>
<td>0.79</td>
<td>0.08</td>
<td>0.15</td>
<td>0.97</td>
<td>2.35</td>
<td>1.47</td>
<td>-1.25</td>
<td>3.13</td>
<td>338</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>3</td>
<td>6</td>
<td>-0.63</td>
<td>-0.59</td>
<td>-0.68</td>
<td>0.93</td>
<td>-0.60</td>
<td>-0.56</td>
<td>-0.57</td>
<td>-0.66</td>
<td>...</td>
<td>-0.57</td>
<td>-0.50</td>
<td>-0.57</td>
<td>-0.49</td>
<td>1.42</td>
<td>1.47</td>
<td>2.10</td>
<td>-0.45</td>
<td>1.54</td>
<td>5</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4</td>
<td>102</td>
<td>0.67</td>
<td>1.58</td>
<td>-0.95</td>
<td>4.57</td>
<td>0.54</td>
<td>2.33</td>
<td>-0.32</td>
<td>2.37</td>
<td>...</td>
<td>-0.96</td>
<td>-1.38</td>
<td>-0.20</td>
<td>-0.82</td>
<td>2.23</td>
<td>-0.84</td>
<td>-0.48</td>
<td>0.67</td>
<td>0.27</td>
<td>101</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5</td>
<td>303</td>
<td>-1.46</td>
<td>0.28</td>
<td>-0.77</td>
<td>-0.73</td>
<td>-0.09</td>
<td>0.75</td>
<td>-0.47</td>
<td>0.86</td>
<td>...</td>
<td>0.62</td>
<td>0.07</td>
<td>0.23</td>
<td>-1.19</td>
<td>0.35</td>
<td>1.74</td>
<td>1.86</td>
<td>-0.57</td>
<td>0.05</td>
<td>485</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">115748</td>
<td>128283</td>
<td>78</td>
<td>-0.62</td>
<td>-0.26</td>
<td>-0.35</td>
<td>0.22</td>
<td>-0.31</td>
<td>0.47</td>
<td>2.37</td>
<td>-1.71</td>
<td>...</td>
<td>-0.71</td>
<td>-1.48</td>
<td>-1.14</td>
<td>0.32</td>
<td>-1.38</td>
<td>0.88</td>
<td>1.70</td>
<td>-1.70</td>
<td>-0.51</td>
<td>77</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">115749</td>
<td>128284</td>
<td>34</td>
<td>-0.89</td>
<td>-0.99</td>
<td>-0.34</td>
<td>0.44</td>
<td>3.55</td>
<td>1.00</td>
<td>-0.57</td>
<td>-0.97</td>
<td>...</td>
<td>-0.78</td>
<td>0.40</td>
<td>-0.58</td>
<td>0.36</td>
<td>-0.35</td>
<td>0.33</td>
<td>0.67</td>
<td>-0.10</td>
<td>0.22</td>
<td>33</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">115750</td>
<td>128285</td>
<td>97</td>
<td>-0.31</td>
<td>-1.78</td>
<td>0.85</td>
<td>-0.02</td>
<td>-0.16</td>
<td>-0.36</td>
<td>-0.55</td>
<td>-1.89</td>
<td>...</td>
<td>-0.69</td>
<td>-0.37</td>
<td>-1.29</td>
<td>0.81</td>
<td>-0.53</td>
<td>-1.23</td>
<td>-0.91</td>
<td>0.84</td>
<td>-0.40</td>
<td>96</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">115751</td>
<td>128286</td>
<td>201</td>
<td>-0.30</td>
<td>-1.96</td>
<td>-1.51</td>
<td>-1.67</td>
<td>-0.96</td>
<td>-0.51</td>
<td>-0.66</td>
<td>-2.36</td>
<td>...</td>
<td>-0.17</td>
<td>-0.58</td>
<td>0.69</td>
<td>-1.59</td>
<td>-0.13</td>
<td>-1.91</td>
<td>-1.36</td>
<td>-1.90</td>
<td>-1.16</td>
<td>202</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">115752</td>
<td>128287</td>
<td>56</td>
<td>0.01</td>
<td>0.87</td>
<td>1.60</td>
<td>0.62</td>
<td>-0.07</td>
<td>-1.08</td>
<td>0.00</td>
<td>-0.70</td>
<td>...</td>
<td>-1.17</td>
<td>0.92</td>
<td>-0.44</td>
<td>-0.83</td>
<td>1.21</td>
<td>0.35</td>
<td>-0.77</td>
<td>0.05</td>
<td>-1.11</td>
<td>55</td>
</tr>
</tbody>
</table>

<p>115753 rows × 1101 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'VirScan-IgG-Zscores sanity checks'</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># missing values in the dataset</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==missing data row counts=="</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(virs_z.isnull().<span class="bu">sum</span>())</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co"># check data types</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==data types=="</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(virs_z.dtypes)</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary statistics for numerical data</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==summary stats=="</span>)</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(virs_z.describe())</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>VirScan-IgG-Zscores sanity checks
==missing data row counts==
id            0
group         0
73_75_S1      0
73_75_S2      0
73_75_S3      0
             ..
82_84_S187    0
82_84_S188    0
82_84_S189    0
82_84_S190    0
input         0
Length: 1101, dtype: int64
============================
==data types==
id              int64
group           int64
73_75_S1      float64
73_75_S2      float64
73_75_S3      float64
               ...   
82_84_S187    float64
82_84_S188    float64
82_84_S189    float64
82_84_S190    float64
input           int64
Length: 1101, dtype: object
============================
==summary stats==
                  id          group       73_75_S1       73_75_S2  \
count  115753.000000  115753.000000  115753.000000  115753.000000   
mean    60060.998194     150.294887       0.446440       0.863865   
std     36635.929238     100.216878      11.789777      15.820201   
min         1.000000       1.000000      -4.630000      -4.250000   
25%     29014.000000      59.000000      -0.810000      -0.820000   
50%     57957.000000     136.000000      -0.090000      -0.090000   
75%     86895.000000     239.000000       0.770000       0.750000   
max    128287.000000     340.000000    1307.240000    1253.070000   

            73_75_S3       73_75_S4       73_75_S5       73_75_S6  \
count  115753.000000  115753.000000  115753.000000  115753.000000   
mean        0.722910       0.513604       0.660813       0.601397   
std        21.370276      10.968330      14.854924      12.268854   
min        -4.270000      -4.550000      -4.160000      -3.910000   
25%        -0.820000      -0.820000      -0.820000      -0.800000   
50%        -0.090000      -0.090000      -0.090000      -0.110000   
75%         0.770000       0.770000       0.760000       0.750000   
max      2376.550000    1028.810000    2021.680000     999.650000   

            73_75_S7       73_75_S8  ...     82_84_S182     82_84_S183  \
count  115753.000000  115753.000000  ...  115753.000000  115753.000000   
mean        1.124989       0.680186  ...       0.704096       0.714590   
std        20.677591      17.288939  ...      15.355436      14.178673   
min        -4.210000      -4.210000  ...      -3.950000      -4.090000   
25%        -0.810000      -0.810000  ...      -0.820000      -0.810000   
50%        -0.110000      -0.080000  ...      -0.100000      -0.130000   
75%         0.750000       0.760000  ...       0.750000       0.720000   
max      1346.850000    1575.390000  ...    2191.610000    1376.130000   

          82_84_S184     82_84_S185     82_84_S186     82_84_S187  \
count  115753.000000  115753.000000  115753.000000  115753.000000   
mean        0.774207       0.749140       0.523308       0.674969   
std        21.748520      19.262465      14.606411      22.391208   
min        -4.130000      -4.310000      -4.140000      -4.080000   
25%        -0.820000      -0.810000      -0.820000      -0.830000   
50%        -0.110000      -0.120000      -0.110000      -0.090000   
75%         0.760000       0.730000       0.740000       0.750000   
max      2688.480000    2892.810000    2247.260000    3029.720000   

          82_84_S188     82_84_S189     82_84_S190          input  
count  115753.000000  115753.000000  115753.000000  115753.000000  
mean        0.704576       1.098997       0.865411     208.131858  
std        13.982901      22.325912      22.394041     418.387223  
min        -3.800000      -3.820000      -4.430000       0.000000  
25%        -0.820000      -0.820000      -0.810000      58.000000  
50%        -0.120000      -0.130000      -0.100000     135.000000  
75%         0.750000       0.720000       0.760000     266.000000  
max      1082.240000    2306.140000    2693.180000   61032.000000  

[8 rows x 1101 columns]
============================</code></pre>
</div>
</div>
<div class="cell" data-tags="[]">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'5. Coronavirus-screen-IgG-Zscores.csv'</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>display(cor_scr_z)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Coronavirus-screen-IgG-Zscores columns: </span><span class="sc">{</span>cor_scr_z<span class="sc">.</span>columns<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>5. Coronavirus-screen-IgG-Zscores.csv
Coronavirus-screen-IgG-Zscores columns: Index(['id', 'group', '73_75_S1', '73_75_S2', '73_75_S3', '73_75_S4',
       '73_75_S5', '73_75_S6', '73_75_S7', '73_75_S8',
       ...
       '82_84_S182', '82_84_S183', '82_84_S184', '82_84_S185', '82_84_S186',
       '82_84_S187', '82_84_S188', '82_84_S189', '82_84_S190', 'input'],
      dtype='object', length=1101)
============================</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">id</th>
<th data-quarto-table-cell-role="th">group</th>
<th data-quarto-table-cell-role="th">73_75_S1</th>
<th data-quarto-table-cell-role="th">73_75_S2</th>
<th data-quarto-table-cell-role="th">73_75_S3</th>
<th data-quarto-table-cell-role="th">73_75_S4</th>
<th data-quarto-table-cell-role="th">73_75_S5</th>
<th data-quarto-table-cell-role="th">73_75_S6</th>
<th data-quarto-table-cell-role="th">73_75_S7</th>
<th data-quarto-table-cell-role="th">73_75_S8</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">82_84_S182</th>
<th data-quarto-table-cell-role="th">82_84_S183</th>
<th data-quarto-table-cell-role="th">82_84_S184</th>
<th data-quarto-table-cell-role="th">82_84_S185</th>
<th data-quarto-table-cell-role="th">82_84_S186</th>
<th data-quarto-table-cell-role="th">82_84_S187</th>
<th data-quarto-table-cell-role="th">82_84_S188</th>
<th data-quarto-table-cell-role="th">82_84_S189</th>
<th data-quarto-table-cell-role="th">82_84_S190</th>
<th data-quarto-table-cell-role="th">input</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0</td>
<td>124</td>
<td>1.45</td>
<td>4.06</td>
<td>1.42</td>
<td>2.90</td>
<td>0.00</td>
<td>0.18</td>
<td>-0.67</td>
<td>0.66</td>
<td>...</td>
<td>1.26</td>
<td>1.79</td>
<td>0.56</td>
<td>0.18</td>
<td>1.72</td>
<td>1.13</td>
<td>-0.32</td>
<td>1.15</td>
<td>0.45</td>
<td>1663</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>1</td>
<td>31</td>
<td>-0.59</td>
<td>1.89</td>
<td>4.63</td>
<td>-0.54</td>
<td>2.32</td>
<td>1.41</td>
<td>1.33</td>
<td>1.53</td>
<td>...</td>
<td>-0.10</td>
<td>0.20</td>
<td>0.25</td>
<td>-0.04</td>
<td>-0.48</td>
<td>0.16</td>
<td>0.19</td>
<td>-1.56</td>
<td>0.66</td>
<td>180</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>2</td>
<td>45</td>
<td>1.66</td>
<td>0.67</td>
<td>0.78</td>
<td>-0.16</td>
<td>2.32</td>
<td>-0.13</td>
<td>0.00</td>
<td>2.73</td>
<td>...</td>
<td>0.02</td>
<td>0.64</td>
<td>-1.76</td>
<td>-0.13</td>
<td>-1.13</td>
<td>-2.03</td>
<td>0.30</td>
<td>1.13</td>
<td>-0.47</td>
<td>414</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>3</td>
<td>24</td>
<td>-2.60</td>
<td>-0.64</td>
<td>1.55</td>
<td>-1.54</td>
<td>-1.81</td>
<td>0.19</td>
<td>-0.80</td>
<td>-1.21</td>
<td>...</td>
<td>-1.04</td>
<td>-1.63</td>
<td>-0.60</td>
<td>-2.07</td>
<td>-0.74</td>
<td>-1.46</td>
<td>-1.00</td>
<td>2.09</td>
<td>-0.54</td>
<td>118</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>4</td>
<td>55</td>
<td>1.53</td>
<td>2.43</td>
<td>2.26</td>
<td>1.75</td>
<td>1.74</td>
<td>3.10</td>
<td>1.02</td>
<td>0.92</td>
<td>...</td>
<td>0.21</td>
<td>0.15</td>
<td>1.75</td>
<td>1.20</td>
<td>2.59</td>
<td>1.50</td>
<td>-0.99</td>
<td>1.57</td>
<td>1.28</td>
<td>844</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">11215</td>
<td>58773</td>
<td>23</td>
<td>2.92</td>
<td>-0.56</td>
<td>1.13</td>
<td>-0.02</td>
<td>0.27</td>
<td>-0.18</td>
<td>-1.13</td>
<td>1.10</td>
<td>...</td>
<td>1.98</td>
<td>0.32</td>
<td>0.42</td>
<td>2.28</td>
<td>0.64</td>
<td>0.97</td>
<td>4.73</td>
<td>-1.18</td>
<td>1.40</td>
<td>108</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11216</td>
<td>58774</td>
<td>4</td>
<td>-0.73</td>
<td>-0.58</td>
<td>-0.46</td>
<td>-0.64</td>
<td>-0.42</td>
<td>-0.39</td>
<td>-0.44</td>
<td>1.00</td>
<td>...</td>
<td>-0.56</td>
<td>-0.51</td>
<td>-0.47</td>
<td>-0.54</td>
<td>-0.41</td>
<td>2.09</td>
<td>-0.44</td>
<td>-0.30</td>
<td>-0.46</td>
<td>3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">11217</td>
<td>58775</td>
<td>2</td>
<td>4.43</td>
<td>-0.32</td>
<td>-0.23</td>
<td>2.82</td>
<td>-0.15</td>
<td>-0.18</td>
<td>-0.31</td>
<td>2.96</td>
<td>...</td>
<td>-0.23</td>
<td>-0.23</td>
<td>-0.21</td>
<td>-0.25</td>
<td>6.48</td>
<td>-0.14</td>
<td>-0.20</td>
<td>-0.07</td>
<td>-0.20</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11218</td>
<td>58776</td>
<td>20</td>
<td>-0.03</td>
<td>0.81</td>
<td>2.46</td>
<td>1.43</td>
<td>0.40</td>
<td>0.27</td>
<td>0.38</td>
<td>0.13</td>
<td>...</td>
<td>-1.76</td>
<td>0.96</td>
<td>1.00</td>
<td>-1.33</td>
<td>3.73</td>
<td>-0.11</td>
<td>2.36</td>
<td>-0.30</td>
<td>0.24</td>
<td>76</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">11219</td>
<td>58777</td>
<td>9</td>
<td>0.58</td>
<td>1.42</td>
<td>-0.85</td>
<td>-0.43</td>
<td>0.71</td>
<td>1.29</td>
<td>-0.87</td>
<td>0.25</td>
<td>...</td>
<td>0.02</td>
<td>2.17</td>
<td>0.08</td>
<td>0.15</td>
<td>-0.72</td>
<td>-0.74</td>
<td>-0.82</td>
<td>-0.77</td>
<td>-1.00</td>
<td>17</td>
</tr>
</tbody>
</table>

<p>11220 rows × 1101 columns</p>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate correlation matrix</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Using DataFrame.drop</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>just_z <span class="op">=</span> cor_scr_z.drop(cor_scr_z.columns[[<span class="dv">1</span>, <span class="dv">2</span>]], axis<span class="op">=</span><span class="dv">1</span>, inplace<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>corr_matrix <span class="op">=</span> just_z.corr()</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">20</span>))</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>sns.heatmap(corr_matrix, annot<span class="op">=</span><span class="va">True</span>, cmap<span class="op">=</span><span class="st">'coolwarm'</span>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="st">'Epitope Binding Signals'</span>)</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>plt.savefig(<span class="st">"Coronavirus-screen-IgG-Zscores-heatmap.png"</span>)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Coronavirus-screen-IgG-Zscores sanity checks'</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co"># missing values in the dataset</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==missing data row counts=="</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cor_scr_z.isnull().<span class="bu">sum</span>())</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="co"># check data types</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==data types=="</span>)</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cor_scr_z.dtypes)</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary statistics for numerical data</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"==summary stats=="</span>)</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(cor_scr_z.describe())</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'============================'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Coronavirus-screen-IgG-Zscores sanity checks
==missing data row counts==
id            0
group         0
73_75_S1      0
73_75_S2      0
73_75_S3      0
             ..
82_84_S187    0
82_84_S188    0
82_84_S189    0
82_84_S190    0
input         0
Length: 1101, dtype: int64
============================
==data types==
id              int64
group           int64
73_75_S1      float64
73_75_S2      float64
73_75_S3      float64
               ...   
82_84_S187    float64
82_84_S188    float64
82_84_S189    float64
82_84_S190    float64
input           int64
Length: 1101, dtype: object
============================
==summary stats==
                 id         group      73_75_S1      73_75_S2      73_75_S3  \
count  11220.000000  11220.000000  11220.000000  11220.000000  11220.000000   
mean   23784.963815     51.407932      0.065491      0.597291      0.216877   
std    25895.636797     37.885679      1.276906      8.580546      3.633929   
min        0.000000      1.000000     -7.340000     -9.730000     -7.900000   
25%     2804.750000     14.000000     -0.800000     -0.750000     -0.730000   
50%     5609.500000     50.000000     -0.190000     -0.210000     -0.180000   
75%    55972.250000     75.000000      0.770000      0.730000      0.740000   
max    58777.000000    139.000000     15.470000    346.950000    172.470000   

           73_75_S4      73_75_S5      73_75_S6      73_75_S7      73_75_S8  \
count  11220.000000  11220.000000  11220.000000  11220.000000  11220.000000   
mean       0.364734      0.264305      0.327391      1.022169      0.075155   
std        7.030516      4.926270      5.713070     17.208482      1.562284   
min       -6.650000     -8.410000     -6.870000     -8.070000     -9.000000   
25%       -0.780000     -0.750000     -0.720000     -0.760000     -0.770000   
50%       -0.130000     -0.150000     -0.180000     -0.170000     -0.180000   
75%        0.730000      0.740000      0.730000      0.710000      0.760000   
max      385.500000    186.830000    307.460000    824.670000     62.860000   

       ...    82_84_S182    82_84_S183    82_84_S184    82_84_S185  \
count  ...  11220.000000  11220.000000  11220.000000  11220.000000   
mean   ...      0.103617      0.150749      0.097601      0.250451   
std    ...      1.812402      2.556526      1.566399      6.232416   
min    ...     -3.920000     -3.260000     -6.270000     -5.840000   
25%    ...     -0.770000     -0.740000     -0.740000     -0.710000   
50%    ...     -0.140000     -0.150000     -0.150000     -0.160000   
75%    ...      0.770000      0.730000      0.750000      0.720000   
max    ...     75.840000    130.200000     37.400000    497.960000   

         82_84_S186    82_84_S187    82_84_S188    82_84_S189    82_84_S190  \
count  11220.000000  11220.000000  11220.000000  11220.000000  11220.000000   
mean       0.088935      0.104156      0.133406      0.307976      0.076866   
std        1.480522      1.689303      2.830787      7.679665      1.391940   
min       -5.620000     -5.060000     -4.760000     -7.870000     -7.300000   
25%       -0.720000     -0.740000     -0.760000     -0.702500     -0.730000   
50%       -0.150000     -0.140000     -0.160000     -0.140000     -0.160000   
75%        0.730000      0.700000      0.710000      0.692500      0.750000   
max       40.000000     68.340000    234.760000    388.480000     20.000000   

              input  
count  11220.000000  
mean     640.944920  
std     1175.797254  
min        0.000000  
25%       42.000000  
50%      483.000000  
75%      964.000000  
max    76350.000000  

[8 rows x 1101 columns]
============================</code></pre>
</div>
</div>
</section>
<section id="questions-for-collaboratorrs" class="level2">
<h2 class="anchored" data-anchor-id="questions-for-collaboratorrs">Questions for collaboratorrs</h2>
<ul>
<li>overview of PhIP-Seq: noisy outputs? relation between detected peptides and epitopes?</li>
<li>Is VirScan a type of PhIP-Seq pipeline?</li>
<li>q: are these salivary proteins known? to what extent (e.g., for all vectors known to cause malaria)?
<ul>
<li>can break down proteins via peptide tiling –&gt; synthetic peptides which can be expressed in phage –&gt; easier for experiments
<ul>
<li>q: are peptides resulting from peptide tiling noisy samples? if so what changes are possible? e.g.&nbsp;insertion, deletion, substitution</li>
</ul></li>
</ul></li>
<li>How does peptide tiling work? coronavirus oligonucleotide library: 56-mer peptides tiling every 28 aa across the proteomes of 10 coronavirus strains… q: what does this mean? schematic? - q: relationship between oligonucleotides and peptides expressed in phages?</li>
<li>How are z scores generated?</li>
<li>what specifically do the outputs look like?
<ul>
<li>Let’s list all the information we want from the outputs:
<ul>
<li>classification of vector – Vectorscan?</li>
<li>exposure history – what would this look like?</li>
<li>anything else?</li>
</ul></li>
<li>e.g., how would the outputs help update VectorScan?</li>
</ul></li>
<li>Relation between peptides and ORFs? And what information helps differentiating vectors?</li>
<li>Does there exist a database we can use to match vectors – peptides? Is this what VectorScan is supposed to do?
<ul>
<li>Progress of VectorScan: is this complete?</li>
</ul></li>
<li>Takeaways from each of Shrock2020 + JHU pipelines – what about these papers are relevant to our goals?
<ul>
<li>Shrock 2020</li>
<li>BEER</li>
<li>Dolphyn</li>
<li>AVARDA</li>
</ul></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sns.histplot(data=data, x='reactivity', hue='COVID_status', element='step', kde=True)</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.title('Distribution of Reactivity Scores')</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.xlabel('Reactivity Score')</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.ylabel('Frequency')</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.show()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sns.boxplot(x='COVID_status', y='reactivity', data=data)</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.title('Reactivity Scores by Patient Group')</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.xlabel('COVID-19 Status')</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.ylabel('Reactivity Score')</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.show()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># from scipy.stats import ttest_ind</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # Assume 'reactivity' is the column with numerical data and 'COVID_status' is the group identifier</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># group1 = data[data['COVID_status'] == 'COVID-19']['reactivity']</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># group2 = data[data['COVID_status'] == 'Pre-COVID']['reactivity']</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="co"># # Perform t-test</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="co"># t_stat, p_val = ttest_ind(group1, group2)</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="co"># print('T-statistic:', t_stat)</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a><span class="co"># print('P-value:', p_val)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># # Calculate correlation matrix</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># corr_matrix = data.corr()</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # Plot using seaborn</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.title('Correlation Matrix of Variables')</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.show()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="examining-all-available-data" class="level2">
<h2 class="anchored" data-anchor-id="examining-all-available-data">Examining all available data</h2>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read all .xlsx files in directory into pd dfs </span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> <span class="st">'./data/Shrock2020_all-data/'</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>xlsx_files <span class="op">=</span> [f <span class="cf">for</span> f <span class="kw">in</span> os.listdir(data_dir) <span class="cf">if</span> f.endswith(<span class="st">'.xlsx'</span>)]</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>total_files <span class="op">=</span> <span class="bu">len</span>(xlsx_files)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>all_dfs <span class="op">=</span> {}</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, f <span class="kw">in</span> <span class="bu">enumerate</span>(xlsx_files):</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"processed </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>total_files<span class="sc">}</span><span class="ss"> files - </span><span class="sc">{</span>i<span class="op">/</span>total_files<span class="op">*</span><span class="dv">100</span><span class="sc">:.2f}</span><span class="ss">% complete"</span>)</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    file_path <span class="op">=</span> os.path.join(data_dir, f)</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>    all_dfs[f] <span class="op">=</span> pd.read_excel(file_path, sheet_name<span class="op">=</span><span class="va">None</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>processed 0/12 files - 0.00% complete
processed 1/12 files - 8.33% complete
processed 2/12 files - 16.67% complete
processed 3/12 files - 25.00% complete
processed 4/12 files - 33.33% complete
processed 5/12 files - 41.67% complete
processed 6/12 files - 50.00% complete
processed 7/12 files - 58.33% complete
processed 8/12 files - 66.67% complete
processed 9/12 files - 75.00% complete
processed 10/12 files - 83.33% complete
processed 11/12 files - 91.67% complete</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>KeyboardInterrupt: </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>all_dfs.keys()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>test_df <span class="op">=</span> all_dfs[<span class="st">'abd4250_table_s3.xlsx'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>test_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>