{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b91f7b9e-3f2e-4ea1-b39f-3755c8848471",
   "metadata": {},
   "source": [
    "# Exploratory Data Analysis of PhIP-Seq data from Shrock 2020\n",
    "- this is a good representation of publicly available PhIP-seq data\n",
    "- link to paper: https://www.science.org/doi/full/10.1126/science.abd4250\n",
    "- for detailed annotations of datasets, see HDSI Grant -- 2024-2027/Published PhIP-seq data/VirScan - Shrock, Science, 2020/Shrock_2020_annotations.md on SharePoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eb561e13-9f10-4903-ba2e-4a72f178c5dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy.stats import ttest_ind\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2c7d833d-93c5-4375-9bad-a85215167f65",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "# five distinct dataframes here\n",
    "data_dir = '../data/Shrock_2020/'\n",
    "pat_meta = pd.read_csv(data_dir + 'Shrock_2020_patient-metadata.csv', encoding='utf-8', engine='python', on_bad_lines='warn') \n",
    "cor_lib = pd.read_csv(data_dir + 'Shrock_2020_Coronavirus-library.csv', encoding='utf-8', engine='python', on_bad_lines='warn')\n",
    "cor_scr_z = pd.read_csv(data_dir + 'Shrock_2020_Coronavirus-screen-IgG-Zscores.csv', encoding='utf-8', engine='python', on_bad_lines='warn') \n",
    "virs_z = pd.read_csv(data_dir + 'Shrock_2020_VirScan-IgG-Zscores.csv', encoding='utf-8', engine='python', on_bad_lines='warn') \n",
    "virs_lib = pd.read_csv(data_dir + 'Shrock_2020_VirScan-library.csv', encoding='utf-8', engine='python', on_bad_lines='warn') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16e1f19f-10ec-4f57-93a1-b0b4bfc2ba61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. patient-metadata\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rep_1</th>\n",
       "      <th>rep_2</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Library</th>\n",
       "      <th>IP reagent</th>\n",
       "      <th>COVID19_status</th>\n",
       "      <th>Hospitalized</th>\n",
       "      <th>Pull down antibody</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>77_79_S23</td>\n",
       "      <td>78_80_S23</td>\n",
       "      <td>BWH_1</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>positive</td>\n",
       "      <td>non-hospitalized</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>73_75_S1</td>\n",
       "      <td>74_76_S1</td>\n",
       "      <td>820-0447-1</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>positive</td>\n",
       "      <td>non-hospitalized</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>73_75_S2</td>\n",
       "      <td>74_76_S2</td>\n",
       "      <td>820-0144-1</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>positive</td>\n",
       "      <td>non-hospitalized</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>73_75_S3</td>\n",
       "      <td>74_76_S3</td>\n",
       "      <td>820-0084-1</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>positive</td>\n",
       "      <td>non-hospitalized</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>73_75_S4</td>\n",
       "      <td>74_76_S4</td>\n",
       "      <td>820-0117-1</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>positive</td>\n",
       "      <td>non-hospitalized</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>81_83_S186</td>\n",
       "      <td>82_84_S186</td>\n",
       "      <td>7/25/19_serum_F</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>81_83_S187</td>\n",
       "      <td>82_84_S187</td>\n",
       "      <td>7/25/19_serum_G</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>81_83_S188</td>\n",
       "      <td>82_84_S188</td>\n",
       "      <td>7/25/19_serum_H</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>81_83_S189</td>\n",
       "      <td>82_84_S189</td>\n",
       "      <td>7/25/19_serum_I</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>81_83_S190</td>\n",
       "      <td>82_84_S190</td>\n",
       "      <td>7/25/19_serum_J</td>\n",
       "      <td>vir3 + CoV 56mer + CoV 20mer</td>\n",
       "      <td>6 ug anti-IgG-biot (Southern Biotech)_25 uL St...</td>\n",
       "      <td>negative</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IgG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          rep_1       rep_2           Sample                       Library  \\\n",
       "0     77_79_S23   78_80_S23            BWH_1  vir3 + CoV 56mer + CoV 20mer   \n",
       "1      73_75_S1    74_76_S1       820-0447-1  vir3 + CoV 56mer + CoV 20mer   \n",
       "2      73_75_S2    74_76_S2       820-0144-1  vir3 + CoV 56mer + CoV 20mer   \n",
       "3      73_75_S3    74_76_S3       820-0084-1  vir3 + CoV 56mer + CoV 20mer   \n",
       "4      73_75_S4    74_76_S4       820-0117-1  vir3 + CoV 56mer + CoV 20mer   \n",
       "..          ...         ...              ...                           ...   \n",
       "417  81_83_S186  82_84_S186  7/25/19_serum_F  vir3 + CoV 56mer + CoV 20mer   \n",
       "418  81_83_S187  82_84_S187  7/25/19_serum_G  vir3 + CoV 56mer + CoV 20mer   \n",
       "419  81_83_S188  82_84_S188  7/25/19_serum_H  vir3 + CoV 56mer + CoV 20mer   \n",
       "420  81_83_S189  82_84_S189  7/25/19_serum_I  vir3 + CoV 56mer + CoV 20mer   \n",
       "421  81_83_S190  82_84_S190  7/25/19_serum_J  vir3 + CoV 56mer + CoV 20mer   \n",
       "\n",
       "                                            IP reagent COVID19_status  \\\n",
       "0    6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   \n",
       "1    6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   \n",
       "2    6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   \n",
       "3    6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   \n",
       "4    6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   \n",
       "..                                                 ...            ...   \n",
       "417  6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       negative   \n",
       "418  6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       negative   \n",
       "419  6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       negative   \n",
       "420  6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       negative   \n",
       "421  6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       negative   \n",
       "\n",
       "         Hospitalized Pull down antibody  \n",
       "0    non-hospitalized                IgG  \n",
       "1    non-hospitalized                IgG  \n",
       "2    non-hospitalized                IgG  \n",
       "3    non-hospitalized                IgG  \n",
       "4    non-hospitalized                IgG  \n",
       "..                ...                ...  \n",
       "417               NaN                IgG  \n",
       "418               NaN                IgG  \n",
       "419               NaN                IgG  \n",
       "420               NaN                IgG  \n",
       "421               NaN                IgG  \n",
       "\n",
       "[422 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient-metadata columns: Index(['rep_1', 'rep_2', 'Sample', 'Library', 'IP reagent', 'COVID19_status',\n",
      "       'Hospitalized', 'Pull down antibody'],\n",
      "      dtype='object')\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print(\"1. patient-metadata\")\n",
    "display(pat_meta)\n",
    "print(f\"patient-metadata columns: {pat_meta.columns}\")\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fc459b9-8836-4f94-be0e-431eb9f88062",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "patient-metadata sanity checks\n",
      "==missing data row counts==\n",
      "rep_1                   0\n",
      "rep_2                   0\n",
      "Sample                  0\n",
      "Library                 0\n",
      "IP reagent              0\n",
      "COVID19_status          0\n",
      "Hospitalized          190\n",
      "Pull down antibody      0\n",
      "dtype: int64\n",
      "============================\n",
      "==data types==\n",
      "rep_1                 object\n",
      "rep_2                 object\n",
      "Sample                object\n",
      "Library               object\n",
      "IP reagent            object\n",
      "COVID19_status        object\n",
      "Hospitalized          object\n",
      "Pull down antibody    object\n",
      "dtype: object\n",
      "============================\n",
      "==summary stats==\n",
      "             rep_1       rep_2           Sample                       Library  \\\n",
      "count          422         422              422                           422   \n",
      "unique         422         422              422                             1   \n",
      "top     81_83_S190  82_84_S190  7/25/19_serum_J  vir3 + CoV 56mer + CoV 20mer   \n",
      "freq             1           1                1                           422   \n",
      "\n",
      "                                               IP reagent COVID19_status  \\\n",
      "count                                                 422            422   \n",
      "unique                                                  1              2   \n",
      "top     6 ug anti-IgG-biot (Southern Biotech)_25 uL St...       positive   \n",
      "freq                                                  422            232   \n",
      "\n",
      "            Hospitalized Pull down antibody  \n",
      "count                232                422  \n",
      "unique                 2                  1  \n",
      "top     non-hospitalized                IgG  \n",
      "freq                 131                422  \n",
      "============================\n",
      "==covid19 status==\n",
      "COVID19_status\n",
      "positive    232\n",
      "negative    190\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print('patient-metadata sanity checks')\n",
    "\n",
    "# missing values in the dataset\n",
    "print(\"==missing data row counts==\")\n",
    "print(pat_meta.isnull().sum())\n",
    "print('============================')\n",
    "# drop rows with missing values\n",
    "# data = data.dropna()\n",
    "\n",
    "# check data types\n",
    "print(\"==data types==\")\n",
    "print(pat_meta.dtypes)\n",
    "print('============================')\n",
    "\n",
    "# Summary statistics for numerical data\n",
    "print(\"==summary stats==\")\n",
    "print(pat_meta.describe())\n",
    "print('============================')\n",
    "\n",
    "# Frequency of categories for COVID-19 status\n",
    "print(\"==covid19 status==\")\n",
    "print(pat_meta['COVID19_status'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb3d165c-9bc1-4d84-9d2f-a4aa6c5bee07",
   "metadata": {},
   "source": [
    "## Questions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8529992c-02d2-47a8-9135-5359fa95cf9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2. VirScan-library\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Aclstr50</th>\n",
       "      <th>Bclstr50</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Gene ontology (GO)</th>\n",
       "      <th>Gene ontology IDs</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>...</th>\n",
       "      <th>Species</th>\n",
       "      <th>Subcellular location</th>\n",
       "      <th>Version (entry)</th>\n",
       "      <th>Version (sequence)</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>oligo</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>peptide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>8652.0</td>\n",
       "      <td>15383.0</td>\n",
       "      <td>A0A126</td>\n",
       "      <td>US5</td>\n",
       "      <td>suppression by virus of host apoptotic process</td>\n",
       "      <td>GO:0019050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>\n",
       "      <td>Glycoprotein J</td>\n",
       "      <td>...</td>\n",
       "      <td>Papiine herpesvirus 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>56</td>\n",
       "      <td>1</td>\n",
       "      <td>aGGAATTCCGCTGCGTatgcgcagcttgctgtttgtggtcggtgct...</td>\n",
       "      <td>Vir2</td>\n",
       "      <td>1</td>\n",
       "      <td>MRSLLFVVGAWVAALVTNLTPDAALASGTTTTAAAGNTSATASPGD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>15383.0</td>\n",
       "      <td>15384.0</td>\n",
       "      <td>A0A126</td>\n",
       "      <td>US5</td>\n",
       "      <td>suppression by virus of host apoptotic process</td>\n",
       "      <td>GO:0019050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>\n",
       "      <td>Glycoprotein J</td>\n",
       "      <td>...</td>\n",
       "      <td>Papiine herpesvirus 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>84</td>\n",
       "      <td>2</td>\n",
       "      <td>aTGAATTCGGAGCGGTactacaaccaccgctgccgcagggaacaca...</td>\n",
       "      <td>Vir2</td>\n",
       "      <td>29</td>\n",
       "      <td>TTTTAAAGNTSATASPGDNATSIDAGSTITAAAPPGHSTPWPALPT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>15384.0</td>\n",
       "      <td>3254.0</td>\n",
       "      <td>A0A126</td>\n",
       "      <td>US5</td>\n",
       "      <td>suppression by virus of host apoptotic process</td>\n",
       "      <td>GO:0019050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>\n",
       "      <td>Glycoprotein J</td>\n",
       "      <td>...</td>\n",
       "      <td>Papiine herpesvirus 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>112</td>\n",
       "      <td>3</td>\n",
       "      <td>aGGAATTCCGCTGCGTattaccgctgccgctcctccaggtcattca...</td>\n",
       "      <td>Vir2</td>\n",
       "      <td>57</td>\n",
       "      <td>ITAAAPPGHSTPWPALPTDLALPLVIGGLCALTLAAMGAGALLHRC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3254.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A126</td>\n",
       "      <td>US5</td>\n",
       "      <td>suppression by virus of host apoptotic process</td>\n",
       "      <td>GO:0019050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>\n",
       "      <td>Glycoprotein J</td>\n",
       "      <td>...</td>\n",
       "      <td>Papiine herpesvirus 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>118</td>\n",
       "      <td>4</td>\n",
       "      <td>aTGAATTCGGAGCGGTttgtgcgccctcacactcgcagcaatgggc...</td>\n",
       "      <td>Vir2</td>\n",
       "      <td>85</td>\n",
       "      <td>LCALTLAAMGAGALLHRCCRRCARRRQNVSSVSA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>32531.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A130</td>\n",
       "      <td>US4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...</td>\n",
       "      <td>Glycoprotein G (Fragment)</td>\n",
       "      <td>...</td>\n",
       "      <td>Papiine herpesvirus 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20</td>\n",
       "      <td>5</td>\n",
       "      <td>aGGAATTCCGCTGCGTcgcgatcgcggcccttctcgctctcgcgtg...</td>\n",
       "      <td>Vir2</td>\n",
       "      <td>1</td>\n",
       "      <td>RDRGPSRSRVRYTRLAASEA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128252</th>\n",
       "      <td>128283</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q2VST0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hydrolase activity, hydrolyzing O-glycosyl com...</td>\n",
       "      <td>GO:0004553; GO:0005975</td>\n",
       "      <td>Ziziphus</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>Allergen Ziz m 1</td>\n",
       "      <td>...</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>224</td>\n",
       "      <td>128283</td>\n",
       "      <td>GGAATTCCGCTGCGTAAAGACTTGACAAAGGATCGCACGCGCCCGT...</td>\n",
       "      <td>IEDB</td>\n",
       "      <td>169</td>\n",
       "      <td>KDLTKDRTRPFYLSAAPKCSAYNDSDAYLWTAVETGLFDFVWVKFY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128253</th>\n",
       "      <td>128284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q2VST0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hydrolase activity, hydrolyzing O-glycosyl com...</td>\n",
       "      <td>GO:0004553; GO:0005975</td>\n",
       "      <td>Ziziphus</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>Allergen Ziz m 1</td>\n",
       "      <td>...</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>252</td>\n",
       "      <td>128284</td>\n",
       "      <td>GGAATTCCGCTGCGTTTATGGACTGCGGTCGAGACTGGCCTTTTTG...</td>\n",
       "      <td>IEDB</td>\n",
       "      <td>197</td>\n",
       "      <td>LWTAVETGLFDFVWVKFYNDTSCQYNNDTAAGLDAFYRSWYDWTVS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128254</th>\n",
       "      <td>128285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q2VST0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hydrolase activity, hydrolyzing O-glycosyl com...</td>\n",
       "      <td>GO:0004553; GO:0005975</td>\n",
       "      <td>Ziziphus</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>Allergen Ziz m 1</td>\n",
       "      <td>...</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>280</td>\n",
       "      <td>128285</td>\n",
       "      <td>GGAATTCCGCTGCGTACAGCGGCTGGGCTGGATGCCTTCTACCGCA...</td>\n",
       "      <td>IEDB</td>\n",
       "      <td>225</td>\n",
       "      <td>TAAGLDAFYRSWYDWTVSLAEGNKLLIGIPASNETDNSPLGGYIPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128255</th>\n",
       "      <td>128286</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q2VST0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hydrolase activity, hydrolyzing O-glycosyl com...</td>\n",
       "      <td>GO:0004553; GO:0005975</td>\n",
       "      <td>Ziziphus</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>Allergen Ziz m 1</td>\n",
       "      <td>...</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>308</td>\n",
       "      <td>128286</td>\n",
       "      <td>GGAATTCCGCTGCGTATCCCTGCCAGCAACGAAACGGATAATAGCC...</td>\n",
       "      <td>IEDB</td>\n",
       "      <td>253</td>\n",
       "      <td>IPASNETDNSPLGGYIPSDVLNDQIVSVIMTSSKFGGVNVWNRYYD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128256</th>\n",
       "      <td>128287</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q2VST0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hydrolase activity, hydrolyzing O-glycosyl com...</td>\n",
       "      <td>GO:0004553; GO:0005975</td>\n",
       "      <td>Ziziphus</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>Allergen Ziz m 1</td>\n",
       "      <td>...</td>\n",
       "      <td>Ziziphus mauritiana</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>330</td>\n",
       "      <td>128287</td>\n",
       "      <td>GGAATTCCGCTGCGTATTATGACAAGTTCAAAATTTGGCGGGGTCA...</td>\n",
       "      <td>IEDB</td>\n",
       "      <td>281</td>\n",
       "      <td>IMTSSKFGGVNVWNRYYDLKTNYSSSIILEYVNSGTKYLPLRTKFM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128257 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  Aclstr50  Bclstr50   Entry Gene names  \\\n",
       "0                0    8652.0   15383.0  A0A126        US5   \n",
       "1                1   15383.0   15384.0  A0A126        US5   \n",
       "2                2   15384.0    3254.0  A0A126        US5   \n",
       "3                3    3254.0       NaN  A0A126        US5   \n",
       "4                4   32531.0       NaN  A0A130        US4   \n",
       "...            ...       ...       ...     ...        ...   \n",
       "128252      128283       NaN       NaN  Q2VST0        NaN   \n",
       "128253      128284       NaN       NaN  Q2VST0        NaN   \n",
       "128254      128285       NaN       NaN  Q2VST0        NaN   \n",
       "128255      128286       NaN       NaN  Q2VST0        NaN   \n",
       "128256      128287       NaN       NaN  Q2VST0        NaN   \n",
       "\n",
       "                                       Gene ontology (GO)  \\\n",
       "0          suppression by virus of host apoptotic process   \n",
       "1          suppression by virus of host apoptotic process   \n",
       "2          suppression by virus of host apoptotic process   \n",
       "3          suppression by virus of host apoptotic process   \n",
       "4                                                     NaN   \n",
       "...                                                   ...   \n",
       "128252  hydrolase activity, hydrolyzing O-glycosyl com...   \n",
       "128253  hydrolase activity, hydrolyzing O-glycosyl com...   \n",
       "128254  hydrolase activity, hydrolyzing O-glycosyl com...   \n",
       "128255  hydrolase activity, hydrolyzing O-glycosyl com...   \n",
       "128256  hydrolase activity, hydrolyzing O-glycosyl com...   \n",
       "\n",
       "             Gene ontology IDs     Genus  \\\n",
       "0                   GO:0019050       NaN   \n",
       "1                   GO:0019050       NaN   \n",
       "2                   GO:0019050       NaN   \n",
       "3                   GO:0019050       NaN   \n",
       "4                          NaN       NaN   \n",
       "...                        ...       ...   \n",
       "128252  GO:0004553; GO:0005975  Ziziphus   \n",
       "128253  GO:0004553; GO:0005975  Ziziphus   \n",
       "128254  GO:0004553; GO:0005975  Ziziphus   \n",
       "128255  GO:0004553; GO:0005975  Ziziphus   \n",
       "128256  GO:0004553; GO:0005975  Ziziphus   \n",
       "\n",
       "                                                 Organism  \\\n",
       "0       Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...   \n",
       "1       Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...   \n",
       "2       Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...   \n",
       "3       Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...   \n",
       "4       Cercopithecine herpesvirus 16 (CeHV-16) (Herpe...   \n",
       "...                                                   ...   \n",
       "128252                                Ziziphus mauritiana   \n",
       "128253                                Ziziphus mauritiana   \n",
       "128254                                Ziziphus mauritiana   \n",
       "128255                                Ziziphus mauritiana   \n",
       "128256                                Ziziphus mauritiana   \n",
       "\n",
       "                    Protein names  ...                Species  \\\n",
       "0                  Glycoprotein J  ...  Papiine herpesvirus 2   \n",
       "1                  Glycoprotein J  ...  Papiine herpesvirus 2   \n",
       "2                  Glycoprotein J  ...  Papiine herpesvirus 2   \n",
       "3                  Glycoprotein J  ...  Papiine herpesvirus 2   \n",
       "4       Glycoprotein G (Fragment)  ...  Papiine herpesvirus 2   \n",
       "...                           ...  ...                    ...   \n",
       "128252           Allergen Ziz m 1  ...    Ziziphus mauritiana   \n",
       "128253           Allergen Ziz m 1  ...    Ziziphus mauritiana   \n",
       "128254           Allergen Ziz m 1  ...    Ziziphus mauritiana   \n",
       "128255           Allergen Ziz m 1  ...    Ziziphus mauritiana   \n",
       "128256           Allergen Ziz m 1  ...    Ziziphus mauritiana   \n",
       "\n",
       "       Subcellular location Version (entry)  Version (sequence)  end      id  \\\n",
       "0                       NaN            13.0                 1.0   56       1   \n",
       "1                       NaN            13.0                 1.0   84       2   \n",
       "2                       NaN            13.0                 1.0  112       3   \n",
       "3                       NaN            13.0                 1.0  118       4   \n",
       "4                       NaN             8.0                 1.0   20       5   \n",
       "...                     ...             ...                 ...  ...     ...   \n",
       "128252                  NaN            42.0                 1.0  224  128283   \n",
       "128253                  NaN            42.0                 1.0  252  128284   \n",
       "128254                  NaN            42.0                 1.0  280  128285   \n",
       "128255                  NaN            42.0                 1.0  308  128286   \n",
       "128256                  NaN            42.0                 1.0  330  128287   \n",
       "\n",
       "                                                    oligo source start  \\\n",
       "0       aGGAATTCCGCTGCGTatgcgcagcttgctgtttgtggtcggtgct...   Vir2     1   \n",
       "1       aTGAATTCGGAGCGGTactacaaccaccgctgccgcagggaacaca...   Vir2    29   \n",
       "2       aGGAATTCCGCTGCGTattaccgctgccgctcctccaggtcattca...   Vir2    57   \n",
       "3       aTGAATTCGGAGCGGTttgtgcgccctcacactcgcagcaatgggc...   Vir2    85   \n",
       "4       aGGAATTCCGCTGCGTcgcgatcgcggcccttctcgctctcgcgtg...   Vir2     1   \n",
       "...                                                   ...    ...   ...   \n",
       "128252  GGAATTCCGCTGCGTAAAGACTTGACAAAGGATCGCACGCGCCCGT...   IEDB   169   \n",
       "128253  GGAATTCCGCTGCGTTTATGGACTGCGGTCGAGACTGGCCTTTTTG...   IEDB   197   \n",
       "128254  GGAATTCCGCTGCGTACAGCGGCTGGGCTGGATGCCTTCTACCGCA...   IEDB   225   \n",
       "128255  GGAATTCCGCTGCGTATCCCTGCCAGCAACGAAACGGATAATAGCC...   IEDB   253   \n",
       "128256  GGAATTCCGCTGCGTATTATGACAAGTTCAAAATTTGGCGGGGTCA...   IEDB   281   \n",
       "\n",
       "                                                  peptide  \n",
       "0       MRSLLFVVGAWVAALVTNLTPDAALASGTTTTAAAGNTSATASPGD...  \n",
       "1       TTTTAAAGNTSATASPGDNATSIDAGSTITAAAPPGHSTPWPALPT...  \n",
       "2       ITAAAPPGHSTPWPALPTDLALPLVIGGLCALTLAAMGAGALLHRC...  \n",
       "3                      LCALTLAAMGAGALLHRCCRRCARRRQNVSSVSA  \n",
       "4                                    RDRGPSRSRVRYTRLAASEA  \n",
       "...                                                   ...  \n",
       "128252  KDLTKDRTRPFYLSAAPKCSAYNDSDAYLWTAVETGLFDFVWVKFY...  \n",
       "128253  LWTAVETGLFDFVWVKFYNDTSCQYNNDTAAGLDAFYRSWYDWTVS...  \n",
       "128254  TAAGLDAFYRSWYDWTVSLAEGNKLLIGIPASNETDNSPLGGYIPS...  \n",
       "128255  IPASNETDNSPLGGYIPSDVLNDQIVSVIMTSSKFGGVNVWNRYYD...  \n",
       "128256  IMTSSKFGGVNVWNRYYDLKTNYSSSIILEYVNSGTKYLPLRTKFM...  \n",
       "\n",
       "[128257 rows x 21 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VirScan-library columns: Index(['Unnamed: 0', 'Aclstr50', 'Bclstr50', 'Entry', 'Gene names',\n",
      "       'Gene ontology (GO)', 'Gene ontology IDs', 'Genus', 'Organism',\n",
      "       'Protein names', 'Sequence', 'Species', 'Subcellular location',\n",
      "       'Version (entry)', 'Version (sequence)', 'end', 'id', 'oligo', 'source',\n",
      "       'start', 'peptide'],\n",
      "      dtype='object')\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print(\"2. VirScan-library\")\n",
    "display(virs_lib)\n",
    "print(f\"VirScan-library columns: {virs_lib.columns}\")\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13892e85-d5aa-48dd-91b7-2a94804b1c15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VirScan-library sanity checks\n",
      "==missing data row counts==\n",
      "Unnamed: 0                   0\n",
      "Aclstr50                 19655\n",
      "Bclstr50                 23874\n",
      "Entry                        0\n",
      "Gene names               43678\n",
      "Gene ontology (GO)       19432\n",
      "Gene ontology IDs        48087\n",
      "Genus                   121852\n",
      "Organism                  2187\n",
      "Protein names                0\n",
      "Sequence                     0\n",
      "Species                      0\n",
      "Subcellular location     75754\n",
      "Version (entry)           2187\n",
      "Version (sequence)        2187\n",
      "end                          0\n",
      "id                           0\n",
      "oligo                        0\n",
      "source                       0\n",
      "start                        0\n",
      "peptide                      0\n",
      "dtype: int64\n",
      "============================\n",
      "==data types==\n",
      "Unnamed: 0                int64\n",
      "Aclstr50                float64\n",
      "Bclstr50                float64\n",
      "Entry                    object\n",
      "Gene names               object\n",
      "Gene ontology (GO)       object\n",
      "Gene ontology IDs        object\n",
      "Genus                    object\n",
      "Organism                 object\n",
      "Protein names            object\n",
      "Sequence                 object\n",
      "Species                  object\n",
      "Subcellular location     object\n",
      "Version (entry)         float64\n",
      "Version (sequence)      float64\n",
      "end                       int64\n",
      "id                        int64\n",
      "oligo                    object\n",
      "source                   object\n",
      "start                     int64\n",
      "peptide                  object\n",
      "dtype: object\n",
      "============================\n",
      "==summary stats==\n",
      "               start            end\n",
      "count  128257.000000  128257.000000\n",
      "mean      542.589387     596.482102\n",
      "std       840.458759     840.761116\n",
      "min         0.000000       6.000000\n",
      "25%        85.000000     140.000000\n",
      "50%       253.000000     308.000000\n",
      "75%       589.000000     644.000000\n",
      "max      7197.000000    7241.000000\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('VirScan-library sanity checks')\n",
    "\n",
    "# missing values in the dataset\n",
    "print(\"==missing data row counts==\")\n",
    "print(virs_lib.isnull().sum())\n",
    "print('============================')\n",
    "\n",
    "# check data types\n",
    "print(\"==data types==\")\n",
    "print(virs_lib.dtypes)\n",
    "print('============================')\n",
    "\n",
    "# Summary statistics for numerical data\n",
    "print(\"==summary stats==\")\n",
    "print(virs_lib[['start', 'end']].describe())\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33358c6c-56b3-4e9a-9d5e-6e6798d3d61c",
   "metadata": {},
   "source": [
    "## Questions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbbdafa1-3611-45f4-8ac4-82d8179c6096",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3. Coronavirus-library\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Library type</th>\n",
       "      <th>Organsim</th>\n",
       "      <th>Barcode ID</th>\n",
       "      <th>Peptide ID</th>\n",
       "      <th>Protein ID</th>\n",
       "      <th>Full protein sequence</th>\n",
       "      <th>Protein length</th>\n",
       "      <th>Protein name</th>\n",
       "      <th>Peptide sequence</th>\n",
       "      <th>Start position</th>\n",
       "      <th>End position</th>\n",
       "      <th>Peptide length</th>\n",
       "      <th>Nucleotide sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Coronavirus 56mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>YP_009725301.9.BC1</td>\n",
       "      <td>YP_009725301.9</td>\n",
       "      <td>YP_009725301</td>\n",
       "      <td>SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...</td>\n",
       "      <td>307</td>\n",
       "      <td>3C-like proteinase</td>\n",
       "      <td>PLSAQTGIAVLDMCASLKELLQNGMNGRTILGSALLEDEFTPFDVV...</td>\n",
       "      <td>251</td>\n",
       "      <td>307</td>\n",
       "      <td>55</td>\n",
       "      <td>GAATTCGGAGCGGTCCTTTGTCAGCACAAACAGGCATTGCCGTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Coronavirus 56mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>YP_009725299.68.BC1</td>\n",
       "      <td>YP_009725299.68</td>\n",
       "      <td>YP_009725299</td>\n",
       "      <td>APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...</td>\n",
       "      <td>1946</td>\n",
       "      <td>nsp3</td>\n",
       "      <td>IALIWNVKDFMSLSEQLRKQIRSAAKKNNLPFKLTCATTRQVVNVV...</td>\n",
       "      <td>1890</td>\n",
       "      <td>1946</td>\n",
       "      <td>55</td>\n",
       "      <td>GAATTCGGAGCGGTATTGCTCTTATTTGGAACGTGAAAGACTTTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Coronavirus 56mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>YP_009725306.3.BC1</td>\n",
       "      <td>YP_009725306.3</td>\n",
       "      <td>YP_009725306</td>\n",
       "      <td>AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...</td>\n",
       "      <td>140</td>\n",
       "      <td>nsp10</td>\n",
       "      <td>NPKGFCDLKGKYVQIPTTCANDPVGFTLKNTVCTVCGMWKGYGCSC...</td>\n",
       "      <td>84</td>\n",
       "      <td>140</td>\n",
       "      <td>55</td>\n",
       "      <td>GAATTCGGAGCGGTAATCCTAAGGGCTTTTGCGACCTCAAGGGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Coronavirus 56mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>YP_009725312.0.BC1</td>\n",
       "      <td>YP_009725312.0</td>\n",
       "      <td>YP_009725312</td>\n",
       "      <td>SADAQSFLNGFAV*</td>\n",
       "      <td>14</td>\n",
       "      <td>nsp11</td>\n",
       "      <td>SADAQSFLNGFAV*</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>GAATTCGGAGCGGTTCCGCTGATGCCCAGTCATTTCTGAACGGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Coronavirus 56mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>YP_009725308.20.BC1</td>\n",
       "      <td>YP_009725308.20</td>\n",
       "      <td>YP_009725308</td>\n",
       "      <td>AVGACVLCNSQTSLRCGACIRRPFLCCKCCYDHVISTSHKLVLSVN...</td>\n",
       "      <td>602</td>\n",
       "      <td>helicase</td>\n",
       "      <td>TQTTETAHSCNVNRFNVAITRAKVGILCIMSDRDLYDKLQFTSLEI...</td>\n",
       "      <td>546</td>\n",
       "      <td>602</td>\n",
       "      <td>55</td>\n",
       "      <td>GAATTCGGAGCGGTACTCAGACAACAGAGACAGCACACTCATGCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58773</th>\n",
       "      <td>58773</td>\n",
       "      <td>SARS-CoV-2 20mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>QID21074.20aa.2.BC2</td>\n",
       "      <td>QID21074.20aa.2</td>\n",
       "      <td>QID21074</td>\n",
       "      <td>MKFLVFLGIIITVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...</td>\n",
       "      <td>122</td>\n",
       "      <td>orf8 protein</td>\n",
       "      <td>ITVAAFHQECSLQSCTQHQP*</td>\n",
       "      <td>10</td>\n",
       "      <td>30</td>\n",
       "      <td>20</td>\n",
       "      <td>GAATTCCGCTGCGTATTACAGTGGCCGCCTTTCACCAGGAGTGCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58774</th>\n",
       "      <td>58774</td>\n",
       "      <td>SARS-CoV-2 20mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>QID21067.20aa.1257.BC2</td>\n",
       "      <td>QID21067.20aa.1257</td>\n",
       "      <td>QID21067</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>7097</td>\n",
       "      <td>orf1ab polyprotein</td>\n",
       "      <td>YKIEELFYSYATHSDKFTAG*</td>\n",
       "      <td>6285</td>\n",
       "      <td>6305</td>\n",
       "      <td>20</td>\n",
       "      <td>GAATTCCGCTGCGTTATAAAATTGAAGAATTGTTTTATTCATACGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58775</th>\n",
       "      <td>58775</td>\n",
       "      <td>SARS-CoV-2 20mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>QID21067.20aa.1258.BC2</td>\n",
       "      <td>QID21067.20aa.1258</td>\n",
       "      <td>QID21067</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>7097</td>\n",
       "      <td>orf1ab polyprotein</td>\n",
       "      <td>LFYSYATHSDKFTAGVCLFW*</td>\n",
       "      <td>6290</td>\n",
       "      <td>6310</td>\n",
       "      <td>20</td>\n",
       "      <td>GAATTCCGCTGCGTTTGTTTTACTCATACGCTACTCATAGTGACAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58776</th>\n",
       "      <td>58776</td>\n",
       "      <td>SARS-CoV-2 20mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>QID21067.20aa.1259.BC2</td>\n",
       "      <td>QID21067.20aa.1259</td>\n",
       "      <td>QID21067</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>7097</td>\n",
       "      <td>orf1ab polyprotein</td>\n",
       "      <td>ATHSDKFTAGVCLFWNCNVD*</td>\n",
       "      <td>6295</td>\n",
       "      <td>6315</td>\n",
       "      <td>20</td>\n",
       "      <td>GAATTCCGCTGCGTGCTACTCACAGCGACAAATTTACAGCCGGTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58777</th>\n",
       "      <td>58777</td>\n",
       "      <td>SARS-CoV-2 20mer</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>QID21067.20aa.1260.BC2</td>\n",
       "      <td>QID21067.20aa.1260</td>\n",
       "      <td>QID21067</td>\n",
       "      <td>MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...</td>\n",
       "      <td>7097</td>\n",
       "      <td>orf1ab polyprotein</td>\n",
       "      <td>KFTAGVCLFWNCNVDRYPAN*</td>\n",
       "      <td>6300</td>\n",
       "      <td>6320</td>\n",
       "      <td>20</td>\n",
       "      <td>GAATTCCGCTGCGTAAGTTCACAGCTGGTGTGTGTCTTTTCTGGAA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58778 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0       Library type  \\\n",
       "0               0  Coronavirus 56mer   \n",
       "1               1  Coronavirus 56mer   \n",
       "2               2  Coronavirus 56mer   \n",
       "3               3  Coronavirus 56mer   \n",
       "4               4  Coronavirus 56mer   \n",
       "...           ...                ...   \n",
       "58773       58773   SARS-CoV-2 20mer   \n",
       "58774       58774   SARS-CoV-2 20mer   \n",
       "58775       58775   SARS-CoV-2 20mer   \n",
       "58776       58776   SARS-CoV-2 20mer   \n",
       "58777       58777   SARS-CoV-2 20mer   \n",
       "\n",
       "                                              Organsim  \\\n",
       "0      Severe acute respiratory syndrome coronavirus 2   \n",
       "1      Severe acute respiratory syndrome coronavirus 2   \n",
       "2      Severe acute respiratory syndrome coronavirus 2   \n",
       "3      Severe acute respiratory syndrome coronavirus 2   \n",
       "4      Severe acute respiratory syndrome coronavirus 2   \n",
       "...                                                ...   \n",
       "58773  Severe acute respiratory syndrome coronavirus 2   \n",
       "58774  Severe acute respiratory syndrome coronavirus 2   \n",
       "58775  Severe acute respiratory syndrome coronavirus 2   \n",
       "58776  Severe acute respiratory syndrome coronavirus 2   \n",
       "58777  Severe acute respiratory syndrome coronavirus 2   \n",
       "\n",
       "                   Barcode ID          Peptide ID    Protein ID  \\\n",
       "0          YP_009725301.9.BC1      YP_009725301.9  YP_009725301   \n",
       "1         YP_009725299.68.BC1     YP_009725299.68  YP_009725299   \n",
       "2          YP_009725306.3.BC1      YP_009725306.3  YP_009725306   \n",
       "3          YP_009725312.0.BC1      YP_009725312.0  YP_009725312   \n",
       "4         YP_009725308.20.BC1     YP_009725308.20  YP_009725308   \n",
       "...                       ...                 ...           ...   \n",
       "58773     QID21074.20aa.2.BC2     QID21074.20aa.2      QID21074   \n",
       "58774  QID21067.20aa.1257.BC2  QID21067.20aa.1257      QID21067   \n",
       "58775  QID21067.20aa.1258.BC2  QID21067.20aa.1258      QID21067   \n",
       "58776  QID21067.20aa.1259.BC2  QID21067.20aa.1259      QID21067   \n",
       "58777  QID21067.20aa.1260.BC2  QID21067.20aa.1260      QID21067   \n",
       "\n",
       "                                   Full protein sequence  Protein length  \\\n",
       "0      SGFRKMAFPSGKVEGCMVQVTCGTTTLNGLWLDDVVYCPRHVICTS...             307   \n",
       "1      APTKVTFGDDTVIEVQGYKSVNITFELDERIDKVLNEKCSAYTVEL...            1946   \n",
       "2      AGNATEVPANSTVLSFCAFAVDAAKAYKDYLASGGQPITNCVKMLC...             140   \n",
       "3                                         SADAQSFLNGFAV*              14   \n",
       "4      AVGACVLCNSQTSLRCGACIRRPFLCCKCCYDHVISTSHKLVLSVN...             602   \n",
       "...                                                  ...             ...   \n",
       "58773  MKFLVFLGIIITVAAFHQECSLQSCTQHQPYVVDDPCPIHFYSKWY...             122   \n",
       "58774  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...            7097   \n",
       "58775  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...            7097   \n",
       "58776  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...            7097   \n",
       "58777  MESLVPGFNEKTHVQLSLPVLQVRDVLVRGFGDSVEEVLSEARQHL...            7097   \n",
       "\n",
       "             Protein name                                   Peptide sequence  \\\n",
       "0      3C-like proteinase  PLSAQTGIAVLDMCASLKELLQNGMNGRTILGSALLEDEFTPFDVV...   \n",
       "1                    nsp3  IALIWNVKDFMSLSEQLRKQIRSAAKKNNLPFKLTCATTRQVVNVV...   \n",
       "2                   nsp10  NPKGFCDLKGKYVQIPTTCANDPVGFTLKNTVCTVCGMWKGYGCSC...   \n",
       "3                   nsp11                                     SADAQSFLNGFAV*   \n",
       "4                helicase  TQTTETAHSCNVNRFNVAITRAKVGILCIMSDRDLYDKLQFTSLEI...   \n",
       "...                   ...                                                ...   \n",
       "58773        orf8 protein                              ITVAAFHQECSLQSCTQHQP*   \n",
       "58774  orf1ab polyprotein                              YKIEELFYSYATHSDKFTAG*   \n",
       "58775  orf1ab polyprotein                              LFYSYATHSDKFTAGVCLFW*   \n",
       "58776  orf1ab polyprotein                              ATHSDKFTAGVCLFWNCNVD*   \n",
       "58777  orf1ab polyprotein                              KFTAGVCLFWNCNVDRYPAN*   \n",
       "\n",
       "       Start position  End position  Peptide length  \\\n",
       "0                 251           307              55   \n",
       "1                1890          1946              55   \n",
       "2                  84           140              55   \n",
       "3                   0            14              13   \n",
       "4                 546           602              55   \n",
       "...               ...           ...             ...   \n",
       "58773              10            30              20   \n",
       "58774            6285          6305              20   \n",
       "58775            6290          6310              20   \n",
       "58776            6295          6315              20   \n",
       "58777            6300          6320              20   \n",
       "\n",
       "                                     Nucleotide sequence  \n",
       "0      GAATTCGGAGCGGTCCTTTGTCAGCACAAACAGGCATTGCCGTGCT...  \n",
       "1      GAATTCGGAGCGGTATTGCTCTTATTTGGAACGTGAAAGACTTTAT...  \n",
       "2      GAATTCGGAGCGGTAATCCTAAGGGCTTTTGCGACCTCAAGGGGAA...  \n",
       "3      GAATTCGGAGCGGTTCCGCTGATGCCCAGTCATTTCTGAACGGTTT...  \n",
       "4      GAATTCGGAGCGGTACTCAGACAACAGAGACAGCACACTCATGCAA...  \n",
       "...                                                  ...  \n",
       "58773  GAATTCCGCTGCGTATTACAGTGGCCGCCTTTCACCAGGAGTGCTC...  \n",
       "58774  GAATTCCGCTGCGTTATAAAATTGAAGAATTGTTTTATTCATACGC...  \n",
       "58775  GAATTCCGCTGCGTTTGTTTTACTCATACGCTACTCATAGTGACAA...  \n",
       "58776  GAATTCCGCTGCGTGCTACTCACAGCGACAAATTTACAGCCGGTGT...  \n",
       "58777  GAATTCCGCTGCGTAAGTTCACAGCTGGTGTGTGTCTTTTCTGGAA...  \n",
       "\n",
       "[58778 rows x 14 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coronavirus-library columns: Index(['Unnamed: 0', 'Library type', 'Organsim', 'Barcode ID', 'Peptide ID',\n",
      "       'Protein ID', 'Full protein sequence', 'Protein length', 'Protein name',\n",
      "       'Peptide sequence', 'Start position', 'End position', 'Peptide length',\n",
      "       'Nucleotide sequence'],\n",
      "      dtype='object')\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('3. Coronavirus-library')\n",
    "display(cor_lib)\n",
    "print(f\"Coronavirus-library columns: {cor_lib.columns}\")\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66af73df-7d98-479d-b5d3-43b629df07f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coronavirus-library sanity checks\n",
      "==missing data row counts==\n",
      "Unnamed: 0               0\n",
      "Library type             0\n",
      "Organsim                 0\n",
      "Barcode ID               0\n",
      "Peptide ID               0\n",
      "Protein ID               0\n",
      "Full protein sequence    0\n",
      "Protein length           0\n",
      "Protein name             0\n",
      "Peptide sequence         0\n",
      "Start position           0\n",
      "End position             0\n",
      "Peptide length           0\n",
      "Nucleotide sequence      0\n",
      "dtype: int64\n",
      "============================\n",
      "==data types==\n",
      "Unnamed: 0                int64\n",
      "Library type             object\n",
      "Organsim                 object\n",
      "Barcode ID               object\n",
      "Peptide ID               object\n",
      "Protein ID               object\n",
      "Full protein sequence    object\n",
      "Protein length            int64\n",
      "Protein name             object\n",
      "Peptide sequence         object\n",
      "Start position            int64\n",
      "End position              int64\n",
      "Peptide length            int64\n",
      "Nucleotide sequence      object\n",
      "dtype: object\n",
      "============================\n",
      "==summary stats==\n",
      "       Protein length  Start position  End position  Peptide length\n",
      "count    58778.000000    58778.000000  58778.000000    58778.000000\n",
      "mean      5078.358161     2492.860288   2546.176427       53.255980\n",
      "std       2930.797729     2238.995872   2238.958572        9.412306\n",
      "min         14.000000        0.000000     14.000000       13.000000\n",
      "25%       1274.000000      364.000000    420.000000       56.000000\n",
      "50%       7097.000000     1932.000000   1988.000000       56.000000\n",
      "75%       7097.000000     4340.000000   4396.000000       56.000000\n",
      "max       7153.000000     7097.000000   7153.000000       56.000000\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('Coronavirus-library sanity checks')\n",
    "\n",
    "# missing values in the dataset\n",
    "print(\"==missing data row counts==\")\n",
    "print(cor_lib.isnull().sum())\n",
    "print('============================')\n",
    "\n",
    "# check data types\n",
    "print(\"==data types==\")\n",
    "print(cor_lib.dtypes)\n",
    "print('============================')\n",
    "\n",
    "# Summary statistics for numerical data\n",
    "print(\"==summary stats==\")\n",
    "print(cor_lib[['Protein length', 'Start position', 'End position', 'Peptide length']].describe())\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3fea50a-2900-444e-8d39-1914a3a91701",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4. VirScan-IgG-Zscores\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>group</th>\n",
       "      <th>73_75_S1</th>\n",
       "      <th>73_75_S2</th>\n",
       "      <th>73_75_S3</th>\n",
       "      <th>73_75_S4</th>\n",
       "      <th>73_75_S5</th>\n",
       "      <th>73_75_S6</th>\n",
       "      <th>73_75_S7</th>\n",
       "      <th>73_75_S8</th>\n",
       "      <th>...</th>\n",
       "      <th>82_84_S182</th>\n",
       "      <th>82_84_S183</th>\n",
       "      <th>82_84_S184</th>\n",
       "      <th>82_84_S185</th>\n",
       "      <th>82_84_S186</th>\n",
       "      <th>82_84_S187</th>\n",
       "      <th>82_84_S188</th>\n",
       "      <th>82_84_S189</th>\n",
       "      <th>82_84_S190</th>\n",
       "      <th>input</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>0.57</td>\n",
       "      <td>-0.95</td>\n",
       "      <td>-0.84</td>\n",
       "      <td>2.74</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>0.70</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>-0.83</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.81</td>\n",
       "      <td>0.61</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>-0.65</td>\n",
       "      <td>1.02</td>\n",
       "      <td>-0.65</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>1.34</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>269</td>\n",
       "      <td>-1.43</td>\n",
       "      <td>0.79</td>\n",
       "      <td>1.35</td>\n",
       "      <td>0.75</td>\n",
       "      <td>2.16</td>\n",
       "      <td>2.66</td>\n",
       "      <td>2.72</td>\n",
       "      <td>-0.48</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.24</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.97</td>\n",
       "      <td>2.35</td>\n",
       "      <td>1.47</td>\n",
       "      <td>-1.25</td>\n",
       "      <td>3.13</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>-0.59</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>0.93</td>\n",
       "      <td>-0.60</td>\n",
       "      <td>-0.56</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-0.50</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-0.49</td>\n",
       "      <td>1.42</td>\n",
       "      <td>1.47</td>\n",
       "      <td>2.10</td>\n",
       "      <td>-0.45</td>\n",
       "      <td>1.54</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>102</td>\n",
       "      <td>0.67</td>\n",
       "      <td>1.58</td>\n",
       "      <td>-0.95</td>\n",
       "      <td>4.57</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2.33</td>\n",
       "      <td>-0.32</td>\n",
       "      <td>2.37</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.96</td>\n",
       "      <td>-1.38</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>-0.82</td>\n",
       "      <td>2.23</td>\n",
       "      <td>-0.84</td>\n",
       "      <td>-0.48</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.27</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>303</td>\n",
       "      <td>-1.46</td>\n",
       "      <td>0.28</td>\n",
       "      <td>-0.77</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>0.75</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>0.86</td>\n",
       "      <td>...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.23</td>\n",
       "      <td>-1.19</td>\n",
       "      <td>0.35</td>\n",
       "      <td>1.74</td>\n",
       "      <td>1.86</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>0.05</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115748</th>\n",
       "      <td>128283</td>\n",
       "      <td>78</td>\n",
       "      <td>-0.62</td>\n",
       "      <td>-0.26</td>\n",
       "      <td>-0.35</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-0.31</td>\n",
       "      <td>0.47</td>\n",
       "      <td>2.37</td>\n",
       "      <td>-1.71</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.71</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>-1.14</td>\n",
       "      <td>0.32</td>\n",
       "      <td>-1.38</td>\n",
       "      <td>0.88</td>\n",
       "      <td>1.70</td>\n",
       "      <td>-1.70</td>\n",
       "      <td>-0.51</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115749</th>\n",
       "      <td>128284</td>\n",
       "      <td>34</td>\n",
       "      <td>-0.89</td>\n",
       "      <td>-0.99</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3.55</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>-0.97</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>0.40</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-0.35</td>\n",
       "      <td>0.33</td>\n",
       "      <td>0.67</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>0.22</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115750</th>\n",
       "      <td>128285</td>\n",
       "      <td>97</td>\n",
       "      <td>-0.31</td>\n",
       "      <td>-1.78</td>\n",
       "      <td>0.85</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>-0.36</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>-1.89</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.69</td>\n",
       "      <td>-0.37</td>\n",
       "      <td>-1.29</td>\n",
       "      <td>0.81</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>-1.23</td>\n",
       "      <td>-0.91</td>\n",
       "      <td>0.84</td>\n",
       "      <td>-0.40</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115751</th>\n",
       "      <td>128286</td>\n",
       "      <td>201</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>-1.51</td>\n",
       "      <td>-1.67</td>\n",
       "      <td>-0.96</td>\n",
       "      <td>-0.51</td>\n",
       "      <td>-0.66</td>\n",
       "      <td>-2.36</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.17</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>0.69</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-0.13</td>\n",
       "      <td>-1.91</td>\n",
       "      <td>-1.36</td>\n",
       "      <td>-1.90</td>\n",
       "      <td>-1.16</td>\n",
       "      <td>202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115752</th>\n",
       "      <td>128287</td>\n",
       "      <td>56</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.87</td>\n",
       "      <td>1.60</td>\n",
       "      <td>0.62</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>-1.08</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.70</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.17</td>\n",
       "      <td>0.92</td>\n",
       "      <td>-0.44</td>\n",
       "      <td>-0.83</td>\n",
       "      <td>1.21</td>\n",
       "      <td>0.35</td>\n",
       "      <td>-0.77</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-1.11</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115753 rows Ã— 1101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id  group  73_75_S1  73_75_S2  73_75_S3  73_75_S4  73_75_S5  \\\n",
       "0            1     10      0.57     -0.95     -0.84      2.74     -0.78   \n",
       "1            2    269     -1.43      0.79      1.35      0.75      2.16   \n",
       "2            3      6     -0.63     -0.59     -0.68      0.93     -0.60   \n",
       "3            4    102      0.67      1.58     -0.95      4.57      0.54   \n",
       "4            5    303     -1.46      0.28     -0.77     -0.73     -0.09   \n",
       "...        ...    ...       ...       ...       ...       ...       ...   \n",
       "115748  128283     78     -0.62     -0.26     -0.35      0.22     -0.31   \n",
       "115749  128284     34     -0.89     -0.99     -0.34      0.44      3.55   \n",
       "115750  128285     97     -0.31     -1.78      0.85     -0.02     -0.16   \n",
       "115751  128286    201     -0.30     -1.96     -1.51     -1.67     -0.96   \n",
       "115752  128287     56      0.01      0.87      1.60      0.62     -0.07   \n",
       "\n",
       "        73_75_S6  73_75_S7  73_75_S8  ...  82_84_S182  82_84_S183  82_84_S184  \\\n",
       "0           0.70     -0.73     -0.83  ...       -0.81        0.61       -0.78   \n",
       "1           2.66      2.72     -0.48  ...       -0.24        0.79        0.08   \n",
       "2          -0.56     -0.57     -0.66  ...       -0.57       -0.50       -0.57   \n",
       "3           2.33     -0.32      2.37  ...       -0.96       -1.38       -0.20   \n",
       "4           0.75     -0.47      0.86  ...        0.62        0.07        0.23   \n",
       "...          ...       ...       ...  ...         ...         ...         ...   \n",
       "115748      0.47      2.37     -1.71  ...       -0.71       -1.48       -1.14   \n",
       "115749      1.00     -0.57     -0.97  ...       -0.78        0.40       -0.58   \n",
       "115750     -0.36     -0.55     -1.89  ...       -0.69       -0.37       -1.29   \n",
       "115751     -0.51     -0.66     -2.36  ...       -0.17       -0.58        0.69   \n",
       "115752     -1.08      0.00     -0.70  ...       -1.17        0.92       -0.44   \n",
       "\n",
       "        82_84_S185  82_84_S186  82_84_S187  82_84_S188  82_84_S189  \\\n",
       "0            -0.65        1.02       -0.65       -0.57        1.34   \n",
       "1             0.15        0.97        2.35        1.47       -1.25   \n",
       "2            -0.49        1.42        1.47        2.10       -0.45   \n",
       "3            -0.82        2.23       -0.84       -0.48        0.67   \n",
       "4            -1.19        0.35        1.74        1.86       -0.57   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "115748        0.32       -1.38        0.88        1.70       -1.70   \n",
       "115749        0.36       -0.35        0.33        0.67       -0.10   \n",
       "115750        0.81       -0.53       -1.23       -0.91        0.84   \n",
       "115751       -1.59       -0.13       -1.91       -1.36       -1.90   \n",
       "115752       -0.83        1.21        0.35       -0.77        0.05   \n",
       "\n",
       "        82_84_S190  input  \n",
       "0            -0.78      9  \n",
       "1             3.13    338  \n",
       "2             1.54      5  \n",
       "3             0.27    101  \n",
       "4             0.05    485  \n",
       "...            ...    ...  \n",
       "115748       -0.51     77  \n",
       "115749        0.22     33  \n",
       "115750       -0.40     96  \n",
       "115751       -1.16    202  \n",
       "115752       -1.11     55  \n",
       "\n",
       "[115753 rows x 1101 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VirScan-IgG-Zscores columns: Index(['id', 'group', '73_75_S1', '73_75_S2', '73_75_S3', '73_75_S4',\n",
      "       '73_75_S5', '73_75_S6', '73_75_S7', '73_75_S8',\n",
      "       ...\n",
      "       '82_84_S182', '82_84_S183', '82_84_S184', '82_84_S185', '82_84_S186',\n",
      "       '82_84_S187', '82_84_S188', '82_84_S189', '82_84_S190', 'input'],\n",
      "      dtype='object', length=1101)\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('4. VirScan-IgG-Zscores')\n",
    "display(virs_z)\n",
    "print(f\"VirScan-IgG-Zscores columns: {virs_z.columns}\")\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0942956-c5aa-41c8-b120-9c7b66f92a6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VirScan-IgG-Zscores sanity checks\n",
      "==missing data row counts==\n",
      "id            0\n",
      "group         0\n",
      "73_75_S1      0\n",
      "73_75_S2      0\n",
      "73_75_S3      0\n",
      "             ..\n",
      "82_84_S187    0\n",
      "82_84_S188    0\n",
      "82_84_S189    0\n",
      "82_84_S190    0\n",
      "input         0\n",
      "Length: 1101, dtype: int64\n",
      "============================\n",
      "==data types==\n",
      "id              int64\n",
      "group           int64\n",
      "73_75_S1      float64\n",
      "73_75_S2      float64\n",
      "73_75_S3      float64\n",
      "               ...   \n",
      "82_84_S187    float64\n",
      "82_84_S188    float64\n",
      "82_84_S189    float64\n",
      "82_84_S190    float64\n",
      "input           int64\n",
      "Length: 1101, dtype: object\n",
      "============================\n",
      "==summary stats==\n",
      "                  id          group       73_75_S1       73_75_S2  \\\n",
      "count  115753.000000  115753.000000  115753.000000  115753.000000   \n",
      "mean    60060.998194     150.294887       0.446440       0.863865   \n",
      "std     36635.929238     100.216878      11.789777      15.820201   \n",
      "min         1.000000       1.000000      -4.630000      -4.250000   \n",
      "25%     29014.000000      59.000000      -0.810000      -0.820000   \n",
      "50%     57957.000000     136.000000      -0.090000      -0.090000   \n",
      "75%     86895.000000     239.000000       0.770000       0.750000   \n",
      "max    128287.000000     340.000000    1307.240000    1253.070000   \n",
      "\n",
      "            73_75_S3       73_75_S4       73_75_S5       73_75_S6  \\\n",
      "count  115753.000000  115753.000000  115753.000000  115753.000000   \n",
      "mean        0.722910       0.513604       0.660813       0.601397   \n",
      "std        21.370276      10.968330      14.854924      12.268854   \n",
      "min        -4.270000      -4.550000      -4.160000      -3.910000   \n",
      "25%        -0.820000      -0.820000      -0.820000      -0.800000   \n",
      "50%        -0.090000      -0.090000      -0.090000      -0.110000   \n",
      "75%         0.770000       0.770000       0.760000       0.750000   \n",
      "max      2376.550000    1028.810000    2021.680000     999.650000   \n",
      "\n",
      "            73_75_S7       73_75_S8  ...     82_84_S182     82_84_S183  \\\n",
      "count  115753.000000  115753.000000  ...  115753.000000  115753.000000   \n",
      "mean        1.124989       0.680186  ...       0.704096       0.714590   \n",
      "std        20.677591      17.288939  ...      15.355436      14.178673   \n",
      "min        -4.210000      -4.210000  ...      -3.950000      -4.090000   \n",
      "25%        -0.810000      -0.810000  ...      -0.820000      -0.810000   \n",
      "50%        -0.110000      -0.080000  ...      -0.100000      -0.130000   \n",
      "75%         0.750000       0.760000  ...       0.750000       0.720000   \n",
      "max      1346.850000    1575.390000  ...    2191.610000    1376.130000   \n",
      "\n",
      "          82_84_S184     82_84_S185     82_84_S186     82_84_S187  \\\n",
      "count  115753.000000  115753.000000  115753.000000  115753.000000   \n",
      "mean        0.774207       0.749140       0.523308       0.674969   \n",
      "std        21.748520      19.262465      14.606411      22.391208   \n",
      "min        -4.130000      -4.310000      -4.140000      -4.080000   \n",
      "25%        -0.820000      -0.810000      -0.820000      -0.830000   \n",
      "50%        -0.110000      -0.120000      -0.110000      -0.090000   \n",
      "75%         0.760000       0.730000       0.740000       0.750000   \n",
      "max      2688.480000    2892.810000    2247.260000    3029.720000   \n",
      "\n",
      "          82_84_S188     82_84_S189     82_84_S190          input  \n",
      "count  115753.000000  115753.000000  115753.000000  115753.000000  \n",
      "mean        0.704576       1.098997       0.865411     208.131858  \n",
      "std        13.982901      22.325912      22.394041     418.387223  \n",
      "min        -3.800000      -3.820000      -4.430000       0.000000  \n",
      "25%        -0.820000      -0.820000      -0.810000      58.000000  \n",
      "50%        -0.120000      -0.130000      -0.100000     135.000000  \n",
      "75%         0.750000       0.720000       0.760000     266.000000  \n",
      "max      1082.240000    2306.140000    2693.180000   61032.000000  \n",
      "\n",
      "[8 rows x 1101 columns]\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('VirScan-IgG-Zscores sanity checks')\n",
    "\n",
    "# missing values in the dataset\n",
    "print(\"==missing data row counts==\")\n",
    "print(virs_z.isnull().sum())\n",
    "print('============================')\n",
    "\n",
    "# check data types\n",
    "print(\"==data types==\")\n",
    "print(virs_z.dtypes)\n",
    "print('============================')\n",
    "\n",
    "# Summary statistics for numerical data\n",
    "print(\"==summary stats==\")\n",
    "print(virs_z.describe())\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f109b9f-a29f-450c-9678-6426b968ca5f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5. Coronavirus-screen-IgG-Zscores.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>group</th>\n",
       "      <th>73_75_S1</th>\n",
       "      <th>73_75_S2</th>\n",
       "      <th>73_75_S3</th>\n",
       "      <th>73_75_S4</th>\n",
       "      <th>73_75_S5</th>\n",
       "      <th>73_75_S6</th>\n",
       "      <th>73_75_S7</th>\n",
       "      <th>73_75_S8</th>\n",
       "      <th>...</th>\n",
       "      <th>82_84_S182</th>\n",
       "      <th>82_84_S183</th>\n",
       "      <th>82_84_S184</th>\n",
       "      <th>82_84_S185</th>\n",
       "      <th>82_84_S186</th>\n",
       "      <th>82_84_S187</th>\n",
       "      <th>82_84_S188</th>\n",
       "      <th>82_84_S189</th>\n",
       "      <th>82_84_S190</th>\n",
       "      <th>input</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>124</td>\n",
       "      <td>1.45</td>\n",
       "      <td>4.06</td>\n",
       "      <td>1.42</td>\n",
       "      <td>2.90</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.18</td>\n",
       "      <td>-0.67</td>\n",
       "      <td>0.66</td>\n",
       "      <td>...</td>\n",
       "      <td>1.26</td>\n",
       "      <td>1.79</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.18</td>\n",
       "      <td>1.72</td>\n",
       "      <td>1.13</td>\n",
       "      <td>-0.32</td>\n",
       "      <td>1.15</td>\n",
       "      <td>0.45</td>\n",
       "      <td>1663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>31</td>\n",
       "      <td>-0.59</td>\n",
       "      <td>1.89</td>\n",
       "      <td>4.63</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>2.32</td>\n",
       "      <td>1.41</td>\n",
       "      <td>1.33</td>\n",
       "      <td>1.53</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-0.48</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.19</td>\n",
       "      <td>-1.56</td>\n",
       "      <td>0.66</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>45</td>\n",
       "      <td>1.66</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.78</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>2.32</td>\n",
       "      <td>-0.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.73</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.64</td>\n",
       "      <td>-1.76</td>\n",
       "      <td>-0.13</td>\n",
       "      <td>-1.13</td>\n",
       "      <td>-2.03</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.13</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>-2.60</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>1.55</td>\n",
       "      <td>-1.54</td>\n",
       "      <td>-1.81</td>\n",
       "      <td>0.19</td>\n",
       "      <td>-0.80</td>\n",
       "      <td>-1.21</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.04</td>\n",
       "      <td>-1.63</td>\n",
       "      <td>-0.60</td>\n",
       "      <td>-2.07</td>\n",
       "      <td>-0.74</td>\n",
       "      <td>-1.46</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>2.09</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>55</td>\n",
       "      <td>1.53</td>\n",
       "      <td>2.43</td>\n",
       "      <td>2.26</td>\n",
       "      <td>1.75</td>\n",
       "      <td>1.74</td>\n",
       "      <td>3.10</td>\n",
       "      <td>1.02</td>\n",
       "      <td>0.92</td>\n",
       "      <td>...</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.15</td>\n",
       "      <td>1.75</td>\n",
       "      <td>1.20</td>\n",
       "      <td>2.59</td>\n",
       "      <td>1.50</td>\n",
       "      <td>-0.99</td>\n",
       "      <td>1.57</td>\n",
       "      <td>1.28</td>\n",
       "      <td>844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11215</th>\n",
       "      <td>58773</td>\n",
       "      <td>23</td>\n",
       "      <td>2.92</td>\n",
       "      <td>-0.56</td>\n",
       "      <td>1.13</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>0.27</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>-1.13</td>\n",
       "      <td>1.10</td>\n",
       "      <td>...</td>\n",
       "      <td>1.98</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.42</td>\n",
       "      <td>2.28</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.97</td>\n",
       "      <td>4.73</td>\n",
       "      <td>-1.18</td>\n",
       "      <td>1.40</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11216</th>\n",
       "      <td>58774</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-0.46</td>\n",
       "      <td>-0.64</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>-0.39</td>\n",
       "      <td>-0.44</td>\n",
       "      <td>1.00</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.56</td>\n",
       "      <td>-0.51</td>\n",
       "      <td>-0.47</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>-0.41</td>\n",
       "      <td>2.09</td>\n",
       "      <td>-0.44</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>-0.46</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11217</th>\n",
       "      <td>58775</td>\n",
       "      <td>2</td>\n",
       "      <td>4.43</td>\n",
       "      <td>-0.32</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>2.82</td>\n",
       "      <td>-0.15</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>-0.31</td>\n",
       "      <td>2.96</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>-0.23</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>-0.25</td>\n",
       "      <td>6.48</td>\n",
       "      <td>-0.14</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11218</th>\n",
       "      <td>58776</td>\n",
       "      <td>20</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>0.81</td>\n",
       "      <td>2.46</td>\n",
       "      <td>1.43</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.13</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.76</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.00</td>\n",
       "      <td>-1.33</td>\n",
       "      <td>3.73</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>2.36</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>0.24</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11219</th>\n",
       "      <td>58777</td>\n",
       "      <td>9</td>\n",
       "      <td>0.58</td>\n",
       "      <td>1.42</td>\n",
       "      <td>-0.85</td>\n",
       "      <td>-0.43</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1.29</td>\n",
       "      <td>-0.87</td>\n",
       "      <td>0.25</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02</td>\n",
       "      <td>2.17</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>-0.74</td>\n",
       "      <td>-0.82</td>\n",
       "      <td>-0.77</td>\n",
       "      <td>-1.00</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11220 rows Ã— 1101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id  group  73_75_S1  73_75_S2  73_75_S3  73_75_S4  73_75_S5  \\\n",
       "0          0    124      1.45      4.06      1.42      2.90      0.00   \n",
       "1          1     31     -0.59      1.89      4.63     -0.54      2.32   \n",
       "2          2     45      1.66      0.67      0.78     -0.16      2.32   \n",
       "3          3     24     -2.60     -0.64      1.55     -1.54     -1.81   \n",
       "4          4     55      1.53      2.43      2.26      1.75      1.74   \n",
       "...      ...    ...       ...       ...       ...       ...       ...   \n",
       "11215  58773     23      2.92     -0.56      1.13     -0.02      0.27   \n",
       "11216  58774      4     -0.73     -0.58     -0.46     -0.64     -0.42   \n",
       "11217  58775      2      4.43     -0.32     -0.23      2.82     -0.15   \n",
       "11218  58776     20     -0.03      0.81      2.46      1.43      0.40   \n",
       "11219  58777      9      0.58      1.42     -0.85     -0.43      0.71   \n",
       "\n",
       "       73_75_S6  73_75_S7  73_75_S8  ...  82_84_S182  82_84_S183  82_84_S184  \\\n",
       "0          0.18     -0.67      0.66  ...        1.26        1.79        0.56   \n",
       "1          1.41      1.33      1.53  ...       -0.10        0.20        0.25   \n",
       "2         -0.13      0.00      2.73  ...        0.02        0.64       -1.76   \n",
       "3          0.19     -0.80     -1.21  ...       -1.04       -1.63       -0.60   \n",
       "4          3.10      1.02      0.92  ...        0.21        0.15        1.75   \n",
       "...         ...       ...       ...  ...         ...         ...         ...   \n",
       "11215     -0.18     -1.13      1.10  ...        1.98        0.32        0.42   \n",
       "11216     -0.39     -0.44      1.00  ...       -0.56       -0.51       -0.47   \n",
       "11217     -0.18     -0.31      2.96  ...       -0.23       -0.23       -0.21   \n",
       "11218      0.27      0.38      0.13  ...       -1.76        0.96        1.00   \n",
       "11219      1.29     -0.87      0.25  ...        0.02        2.17        0.08   \n",
       "\n",
       "       82_84_S185  82_84_S186  82_84_S187  82_84_S188  82_84_S189  82_84_S190  \\\n",
       "0            0.18        1.72        1.13       -0.32        1.15        0.45   \n",
       "1           -0.04       -0.48        0.16        0.19       -1.56        0.66   \n",
       "2           -0.13       -1.13       -2.03        0.30        1.13       -0.47   \n",
       "3           -2.07       -0.74       -1.46       -1.00        2.09       -0.54   \n",
       "4            1.20        2.59        1.50       -0.99        1.57        1.28   \n",
       "...           ...         ...         ...         ...         ...         ...   \n",
       "11215        2.28        0.64        0.97        4.73       -1.18        1.40   \n",
       "11216       -0.54       -0.41        2.09       -0.44       -0.30       -0.46   \n",
       "11217       -0.25        6.48       -0.14       -0.20       -0.07       -0.20   \n",
       "11218       -1.33        3.73       -0.11        2.36       -0.30        0.24   \n",
       "11219        0.15       -0.72       -0.74       -0.82       -0.77       -1.00   \n",
       "\n",
       "       input  \n",
       "0       1663  \n",
       "1        180  \n",
       "2        414  \n",
       "3        118  \n",
       "4        844  \n",
       "...      ...  \n",
       "11215    108  \n",
       "11216      3  \n",
       "11217      1  \n",
       "11218     76  \n",
       "11219     17  \n",
       "\n",
       "[11220 rows x 1101 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coronavirus-screen-IgG-Zscores columns: Index(['id', 'group', '73_75_S1', '73_75_S2', '73_75_S3', '73_75_S4',\n",
      "       '73_75_S5', '73_75_S6', '73_75_S7', '73_75_S8',\n",
      "       ...\n",
      "       '82_84_S182', '82_84_S183', '82_84_S184', '82_84_S185', '82_84_S186',\n",
      "       '82_84_S187', '82_84_S188', '82_84_S189', '82_84_S190', 'input'],\n",
      "      dtype='object', length=1101)\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('5. Coronavirus-screen-IgG-Zscores.csv')\n",
    "display(cor_scr_z)\n",
    "print(f\"Coronavirus-screen-IgG-Zscores columns: {cor_scr_z.columns}\")\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03a1bc60-961c-425e-9d85-22e6e6447073",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e017ba-cf4d-4e42-828b-64edaa2e0b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate correlation matrix\n",
    "# Using DataFrame.drop\n",
    "just_z = cor_scr_z.drop(cor_scr_z.columns[[1, 2]], axis=1, inplace=False)\n",
    "corr_matrix = just_z.corr()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,20))\n",
    "sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')\n",
    "plt.title('Epitope Binding Signals')\n",
    "plt.savefig(\"Coronavirus-screen-IgG-Zscores-heatmap.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9b4303c-8d82-4f69-841e-d6282c2ed491",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coronavirus-screen-IgG-Zscores sanity checks\n",
      "==missing data row counts==\n",
      "id            0\n",
      "group         0\n",
      "73_75_S1      0\n",
      "73_75_S2      0\n",
      "73_75_S3      0\n",
      "             ..\n",
      "82_84_S187    0\n",
      "82_84_S188    0\n",
      "82_84_S189    0\n",
      "82_84_S190    0\n",
      "input         0\n",
      "Length: 1101, dtype: int64\n",
      "============================\n",
      "==data types==\n",
      "id              int64\n",
      "group           int64\n",
      "73_75_S1      float64\n",
      "73_75_S2      float64\n",
      "73_75_S3      float64\n",
      "               ...   \n",
      "82_84_S187    float64\n",
      "82_84_S188    float64\n",
      "82_84_S189    float64\n",
      "82_84_S190    float64\n",
      "input           int64\n",
      "Length: 1101, dtype: object\n",
      "============================\n",
      "==summary stats==\n",
      "                 id         group      73_75_S1      73_75_S2      73_75_S3  \\\n",
      "count  11220.000000  11220.000000  11220.000000  11220.000000  11220.000000   \n",
      "mean   23784.963815     51.407932      0.065491      0.597291      0.216877   \n",
      "std    25895.636797     37.885679      1.276906      8.580546      3.633929   \n",
      "min        0.000000      1.000000     -7.340000     -9.730000     -7.900000   \n",
      "25%     2804.750000     14.000000     -0.800000     -0.750000     -0.730000   \n",
      "50%     5609.500000     50.000000     -0.190000     -0.210000     -0.180000   \n",
      "75%    55972.250000     75.000000      0.770000      0.730000      0.740000   \n",
      "max    58777.000000    139.000000     15.470000    346.950000    172.470000   \n",
      "\n",
      "           73_75_S4      73_75_S5      73_75_S6      73_75_S7      73_75_S8  \\\n",
      "count  11220.000000  11220.000000  11220.000000  11220.000000  11220.000000   \n",
      "mean       0.364734      0.264305      0.327391      1.022169      0.075155   \n",
      "std        7.030516      4.926270      5.713070     17.208482      1.562284   \n",
      "min       -6.650000     -8.410000     -6.870000     -8.070000     -9.000000   \n",
      "25%       -0.780000     -0.750000     -0.720000     -0.760000     -0.770000   \n",
      "50%       -0.130000     -0.150000     -0.180000     -0.170000     -0.180000   \n",
      "75%        0.730000      0.740000      0.730000      0.710000      0.760000   \n",
      "max      385.500000    186.830000    307.460000    824.670000     62.860000   \n",
      "\n",
      "       ...    82_84_S182    82_84_S183    82_84_S184    82_84_S185  \\\n",
      "count  ...  11220.000000  11220.000000  11220.000000  11220.000000   \n",
      "mean   ...      0.103617      0.150749      0.097601      0.250451   \n",
      "std    ...      1.812402      2.556526      1.566399      6.232416   \n",
      "min    ...     -3.920000     -3.260000     -6.270000     -5.840000   \n",
      "25%    ...     -0.770000     -0.740000     -0.740000     -0.710000   \n",
      "50%    ...     -0.140000     -0.150000     -0.150000     -0.160000   \n",
      "75%    ...      0.770000      0.730000      0.750000      0.720000   \n",
      "max    ...     75.840000    130.200000     37.400000    497.960000   \n",
      "\n",
      "         82_84_S186    82_84_S187    82_84_S188    82_84_S189    82_84_S190  \\\n",
      "count  11220.000000  11220.000000  11220.000000  11220.000000  11220.000000   \n",
      "mean       0.088935      0.104156      0.133406      0.307976      0.076866   \n",
      "std        1.480522      1.689303      2.830787      7.679665      1.391940   \n",
      "min       -5.620000     -5.060000     -4.760000     -7.870000     -7.300000   \n",
      "25%       -0.720000     -0.740000     -0.760000     -0.702500     -0.730000   \n",
      "50%       -0.150000     -0.140000     -0.160000     -0.140000     -0.160000   \n",
      "75%        0.730000      0.700000      0.710000      0.692500      0.750000   \n",
      "max       40.000000     68.340000    234.760000    388.480000     20.000000   \n",
      "\n",
      "              input  \n",
      "count  11220.000000  \n",
      "mean     640.944920  \n",
      "std     1175.797254  \n",
      "min        0.000000  \n",
      "25%       42.000000  \n",
      "50%      483.000000  \n",
      "75%      964.000000  \n",
      "max    76350.000000  \n",
      "\n",
      "[8 rows x 1101 columns]\n",
      "============================\n"
     ]
    }
   ],
   "source": [
    "print('Coronavirus-screen-IgG-Zscores sanity checks')\n",
    "\n",
    "# missing values in the dataset\n",
    "print(\"==missing data row counts==\")\n",
    "print(cor_scr_z.isnull().sum())\n",
    "print('============================')\n",
    "\n",
    "# check data types\n",
    "print(\"==data types==\")\n",
    "print(cor_scr_z.dtypes)\n",
    "print('============================')\n",
    "\n",
    "# Summary statistics for numerical data\n",
    "print(\"==summary stats==\")\n",
    "print(cor_scr_z.describe())\n",
    "print('============================')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3030c252-efd8-4b87-8622-f81ee8e53403",
   "metadata": {},
   "source": [
    "## Questions for collaboratorrs\n",
    "\n",
    "- overview of PhIP-Seq: noisy outputs? relation between detected peptides and epitopes? \n",
    "- Is VirScan a type of PhIP-Seq pipeline? \n",
    "- q: are these salivary proteins known? to what extent (e.g., for all vectors known to cause malaria)? \n",
    "\t- can break down proteins via peptide tiling --> synthetic peptides which can be expressed in phage --> easier for experiments \n",
    "\t\t- q: are peptides resulting from peptide tiling noisy samples? if so what changes are possible? e.g. insertion, deletion, substitution \n",
    "- How does peptide tiling work? coronavirus oligonucleotide library: 56-mer peptides tiling every 28 aa across the proteomes of 10 coronavirus strains... q: what does this mean? schematic? \n",
    "\t\t- q: relationship between oligonucleotides and peptides expressed in phages? \n",
    "- How are z scores generated?\n",
    "- what specifically do the outputs look like? \n",
    "\t- Let's list all the information we want from the outputs:  \n",
    "\t\t- classification of vector -- Vectorscan? \n",
    "\t\t- exposure history -- what would this look like? \n",
    "\t\t- anything else? \n",
    "\t- e.g., how would the outputs help update VectorScan? \n",
    "- Relation between peptides and ORFs? And what information helps differentiating vectors? \n",
    "- Does there exist a database we can use to match vectors -- peptides? Is this what VectorScan is supposed to do? \n",
    "\t- Progress of VectorScan: is this complete? \n",
    "- Takeaways from each of Shrock2020 + JHU pipelines -- what about these papers are relevant to our goals? \n",
    "\t- Shrock 2020\n",
    "\t- BEER\n",
    "\t- Dolphyn\n",
    "\t- AVARDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e850a438-5f70-4b92-a3fc-3962681c5251",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.histplot(data=data, x='reactivity', hue='COVID_status', element='step', kde=True)\n",
    "# plt.title('Distribution of Reactivity Scores')\n",
    "# plt.xlabel('Reactivity Score')\n",
    "# plt.ylabel('Frequency')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75044fa9-7dbf-4353-984a-ce3ee2cb08ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sns.boxplot(x='COVID_status', y='reactivity', data=data)\n",
    "# plt.title('Reactivity Scores by Patient Group')\n",
    "# plt.xlabel('COVID-19 Status')\n",
    "# plt.ylabel('Reactivity Score')\n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32493177-c765-4088-9e31-2d622a9b5df9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from scipy.stats import ttest_ind\n",
    "\n",
    "# # Assume 'reactivity' is the column with numerical data and 'COVID_status' is the group identifier\n",
    "# group1 = data[data['COVID_status'] == 'COVID-19']['reactivity']\n",
    "# group2 = data[data['COVID_status'] == 'Pre-COVID']['reactivity']\n",
    "\n",
    "# # Perform t-test\n",
    "# t_stat, p_val = ttest_ind(group1, group2)\n",
    "# print('T-statistic:', t_stat)\n",
    "# print('P-value:', p_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30efe6df-6031-421e-a086-1df5ed2537c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Calculate correlation matrix\n",
    "# corr_matrix = data.corr()\n",
    "\n",
    "# # Plot using seaborn\n",
    "# sns.heatmap(corr_matrix, annot=True, cmap='coolwarm')\n",
    "# plt.title('Correlation Matrix of Variables')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db83e45d-2e09-476f-a669-d00b181b1c9d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Examining all available data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e10a576a-a4e1-4194-845f-dd40b8e5d4e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processed 0/12 files - 0.00% complete\n",
      "processed 1/12 files - 8.33% complete\n",
      "processed 2/12 files - 16.67% complete\n",
      "processed 3/12 files - 25.00% complete\n",
      "processed 4/12 files - 33.33% complete\n",
      "processed 5/12 files - 41.67% complete\n",
      "processed 6/12 files - 50.00% complete\n",
      "processed 7/12 files - 58.33% complete\n",
      "processed 8/12 files - 66.67% complete\n",
      "processed 9/12 files - 75.00% complete\n",
      "processed 10/12 files - 83.33% complete\n",
      "processed 11/12 files - 91.67% complete\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 11\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mprocessed \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m/\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtotal_files\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m files - \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;241m/\u001b[39mtotal_files\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m100\u001b[39m\u001b[38;5;132;01m:\u001b[39;00m\u001b[38;5;124m.2f\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m% complete\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     10\u001b[0m file_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(data_dir, f)\n\u001b[0;32m---> 11\u001b[0m all_dfs[f] \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_excel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfile_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msheet_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/site-packages/pandas/io/excel/_base.py:508\u001b[0m, in \u001b[0;36mread_excel\u001b[0;34m(io, sheet_name, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, parse_dates, date_parser, date_format, thousands, decimal, comment, skipfooter, storage_options, dtype_backend, engine_kwargs)\u001b[0m\n\u001b[1;32m    502\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    503\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mEngine should not be specified when passing \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    504\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124man ExcelFile - ExcelFile already has the engine set\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    505\u001b[0m     )\n\u001b[1;32m    507\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 508\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    509\u001b[0m \u001b[43m        \u001b[49m\u001b[43msheet_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msheet_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    510\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheader\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    511\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnames\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    512\u001b[0m \u001b[43m        \u001b[49m\u001b[43mindex_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    513\u001b[0m \u001b[43m        \u001b[49m\u001b[43musecols\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43musecols\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    514\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    515\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconverters\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconverters\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    516\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtrue_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrue_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    517\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfalse_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfalse_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    518\u001b[0m \u001b[43m        \u001b[49m\u001b[43mskiprows\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskiprows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    519\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnrows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    520\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    521\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkeep_default_na\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeep_default_na\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    522\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_filter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_filter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    523\u001b[0m \u001b[43m        \u001b[49m\u001b[43mverbose\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    524\u001b[0m \u001b[43m        \u001b[49m\u001b[43mparse_dates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparse_dates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    525\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdate_parser\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_parser\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    526\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    527\u001b[0m \u001b[43m        \u001b[49m\u001b[43mthousands\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mthousands\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    528\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdecimal\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdecimal\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    529\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcomment\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcomment\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    530\u001b[0m \u001b[43m        \u001b[49m\u001b[43mskipfooter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskipfooter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    531\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdtype_backend\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype_backend\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    532\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    533\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    534\u001b[0m     \u001b[38;5;66;03m# make sure to close opened file handles\u001b[39;00m\n\u001b[1;32m    535\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m should_close:\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/site-packages/pandas/io/excel/_base.py:1616\u001b[0m, in \u001b[0;36mExcelFile.parse\u001b[0;34m(self, sheet_name, header, names, index_col, usecols, converters, true_values, false_values, skiprows, nrows, na_values, parse_dates, date_parser, date_format, thousands, comment, skipfooter, dtype_backend, **kwds)\u001b[0m\n\u001b[1;32m   1576\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mparse\u001b[39m(\n\u001b[1;32m   1577\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m   1578\u001b[0m     sheet_name: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mint\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mint\u001b[39m] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mstr\u001b[39m] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1596\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwds,\n\u001b[1;32m   1597\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DataFrame \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mdict\u001b[39m[\u001b[38;5;28mstr\u001b[39m, DataFrame] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mdict\u001b[39m[\u001b[38;5;28mint\u001b[39m, DataFrame]:\n\u001b[1;32m   1598\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   1599\u001b[0m \u001b[38;5;124;03m    Parse specified sheet(s) into a DataFrame.\u001b[39;00m\n\u001b[1;32m   1600\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1614\u001b[0m \u001b[38;5;124;03m    >>> file.parse()  # doctest: +SKIP\u001b[39;00m\n\u001b[1;32m   1615\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1616\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_reader\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1617\u001b[0m \u001b[43m        \u001b[49m\u001b[43msheet_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msheet_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1618\u001b[0m \u001b[43m        \u001b[49m\u001b[43mheader\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheader\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1619\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnames\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnames\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1620\u001b[0m \u001b[43m        \u001b[49m\u001b[43mindex_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_col\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1621\u001b[0m \u001b[43m        \u001b[49m\u001b[43musecols\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43musecols\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1622\u001b[0m \u001b[43m        \u001b[49m\u001b[43mconverters\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconverters\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1623\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtrue_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtrue_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1624\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfalse_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfalse_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1625\u001b[0m \u001b[43m        \u001b[49m\u001b[43mskiprows\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskiprows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1626\u001b[0m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnrows\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1627\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_values\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_values\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1628\u001b[0m \u001b[43m        \u001b[49m\u001b[43mparse_dates\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparse_dates\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1629\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdate_parser\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_parser\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1630\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1631\u001b[0m \u001b[43m        \u001b[49m\u001b[43mthousands\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mthousands\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1632\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcomment\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcomment\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1633\u001b[0m \u001b[43m        \u001b[49m\u001b[43mskipfooter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mskipfooter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1634\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdtype_backend\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdtype_backend\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1635\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1636\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/site-packages/pandas/io/excel/_base.py:778\u001b[0m, in \u001b[0;36mBaseExcelReader.parse\u001b[0;34m(self, sheet_name, header, names, index_col, usecols, dtype, true_values, false_values, skiprows, nrows, na_values, verbose, parse_dates, date_parser, date_format, thousands, decimal, comment, skipfooter, dtype_backend, **kwds)\u001b[0m\n\u001b[1;32m    775\u001b[0m     sheet \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_sheet_by_index(asheetname)\n\u001b[1;32m    777\u001b[0m file_rows_needed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_calc_rows(header, index_col, skiprows, nrows)\n\u001b[0;32m--> 778\u001b[0m data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_sheet_data\u001b[49m\u001b[43m(\u001b[49m\u001b[43msheet\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfile_rows_needed\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    779\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(sheet, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mclose\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m    780\u001b[0m     \u001b[38;5;66;03m# pyxlsb opens two TemporaryFiles\u001b[39;00m\n\u001b[1;32m    781\u001b[0m     sheet\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/site-packages/pandas/io/excel/_openpyxl.py:615\u001b[0m, in \u001b[0;36mOpenpyxlReader.get_sheet_data\u001b[0;34m(self, sheet, file_rows_needed)\u001b[0m\n\u001b[1;32m    613\u001b[0m data: \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mlist\u001b[39m[Scalar]] \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    614\u001b[0m last_row_with_data \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m\n\u001b[0;32m--> 615\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m row_number, row \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(sheet\u001b[38;5;241m.\u001b[39mrows):\n\u001b[1;32m    616\u001b[0m     converted_row \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_convert_cell(cell) \u001b[38;5;28;01mfor\u001b[39;00m cell \u001b[38;5;129;01min\u001b[39;00m row]\n\u001b[1;32m    617\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m converted_row \u001b[38;5;129;01mand\u001b[39;00m converted_row[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    618\u001b[0m         \u001b[38;5;66;03m# trim trailing empty elements\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/site-packages/openpyxl/worksheet/_read_only.py:81\u001b[0m, in \u001b[0;36mReadOnlyWorksheet._cells_by_row\u001b[0;34m(self, min_col, min_row, max_col, max_row, values_only)\u001b[0m\n\u001b[1;32m     77\u001b[0m src \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_source()\n\u001b[1;32m     78\u001b[0m parser \u001b[38;5;241m=\u001b[39m WorkSheetParser(src, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_shared_strings,\n\u001b[1;32m     79\u001b[0m                          data_only\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparent\u001b[38;5;241m.\u001b[39mdata_only, epoch\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparent\u001b[38;5;241m.\u001b[39mepoch,\n\u001b[1;32m     80\u001b[0m                          date_formats\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mparent\u001b[38;5;241m.\u001b[39m_date_formats)\n\u001b[0;32m---> 81\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m idx, row \u001b[38;5;129;01min\u001b[39;00m parser\u001b[38;5;241m.\u001b[39mparse():\n\u001b[1;32m     82\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m max_row \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m idx \u001b[38;5;241m>\u001b[39m max_row:\n\u001b[1;32m     83\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:156\u001b[0m, in \u001b[0;36mparse\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/xml/etree/ElementTree.py:1258\u001b[0m, in \u001b[0;36miterparse.<locals>.iterator\u001b[0;34m(source)\u001b[0m\n\u001b[1;32m   1256\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m data:\n\u001b[1;32m   1257\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[0;32m-> 1258\u001b[0m     \u001b[43mpullparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfeed\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1259\u001b[0m root \u001b[38;5;241m=\u001b[39m pullparser\u001b[38;5;241m.\u001b[39m_close_and_return_root()\n\u001b[1;32m   1260\u001b[0m \u001b[38;5;28;01myield from\u001b[39;00m pullparser\u001b[38;5;241m.\u001b[39mread_events()\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/xml/etree/ElementTree.py:1296\u001b[0m, in \u001b[0;36mXMLPullParser.feed\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m   1294\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m data:\n\u001b[1;32m   1295\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1296\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_parser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfeed\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1297\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mSyntaxError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m   1298\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_events_queue\u001b[38;5;241m.\u001b[39mappend(exc)\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/xml/etree/ElementTree.py:1718\u001b[0m, in \u001b[0;36mXMLParser.feed\u001b[0;34m(self, data)\u001b[0m\n\u001b[1;32m   1716\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Feed encoded data to parser.\"\"\"\u001b[39;00m\n\u001b[1;32m   1717\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1718\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparser\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mParse\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m   1719\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_error \u001b[38;5;28;01mas\u001b[39;00m v:\n\u001b[1;32m   1720\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raiseerror(v)\n",
      "File \u001b[0;32m/private/var/folders/nz/j6p8yfhx1mv_0grj5xl4650h0000gp/T/abs_9afkyeq1z4/croot/python-split_1715024112201/work/Modules/pyexpat.c:423\u001b[0m, in \u001b[0;36mStartElement\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/llm/lib/python3.10/xml/etree/ElementTree.py:1656\u001b[0m, in \u001b[0;36mXMLParser._start\u001b[0;34m(self, tag, attr_list)\u001b[0m\n\u001b[1;32m   1654\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;28mlen\u001b[39m(attr_list), \u001b[38;5;241m2\u001b[39m):\n\u001b[1;32m   1655\u001b[0m         attrib[fixname(attr_list[i])] \u001b[38;5;241m=\u001b[39m attr_list[i\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m-> 1656\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtarget\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstart\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtag\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mattrib\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# read all .xlsx files in directory into pd dfs \n",
    "data_dir = './data/Shrock2020_all-data/'\n",
    "xlsx_files = [f for f in os.listdir(data_dir) if f.endswith('.xlsx')]\n",
    "\n",
    "total_files = len(xlsx_files)\n",
    "\n",
    "all_dfs = {}\n",
    "for i, f in enumerate(xlsx_files):\n",
    "    print(f\"processed {i}/{total_files} files - {i/total_files*100:.2f}% complete\")\n",
    "    file_path = os.path.join(data_dir, f)\n",
    "    all_dfs[f] = pd.read_excel(file_path, sheet_name=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58cc816f",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da3c688-c5a7-41cb-aff2-2f82c456ba87",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_dfs.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c97e6aed-da32-440b-ade9-6d8ad266ae28",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df = all_dfs['abd4250_table_s3.xlsx']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c90cb9bf-65b1-4860-8a6d-66b1a49db014",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
